// angular-cesium/esm5/lib/angular-cesium/services/drawers/entities-drawer/optimized-entity-collection.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var OptimizedEntityCollection = /** @class */ (function () {
    function OptimizedEntityCollection(entityCollection, collectionSize, updateRate) {
        if (collectionSize === void 0) { collectionSize = -1; }
        if (updateRate === void 0) { updateRate = -1; }
        this.entityCollection = entityCollection;
        this._isSuspended = false;
        this._isHardSuspend = false;
        this._updateRate = updateRate;
        this._collectionSize = collectionSize;
    }
    /**
     * @param {?} show
     * @return {?}
     */
    OptimizedEntityCollection.prototype.setShow = /**
     * @param {?} show
     * @return {?}
     */
    function (show) {
        this.entityCollection.show = show;
    };
    Object.defineProperty(OptimizedEntityCollection.prototype, "isSuspended", {
        get: /**
         * @return {?}
         */
        function () {
            return this._isSuspended;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(OptimizedEntityCollection.prototype, "updateRate", {
        get: /**
         * @return {?}
         */
        function () {
            return this._updateRate;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._updateRate = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(OptimizedEntityCollection.prototype, "collectionSize", {
        get: /**
         * @return {?}
         */
        function () {
            return this._collectionSize;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._collectionSize = value;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    OptimizedEntityCollection.prototype.collection = /**
     * @return {?}
     */
    function () {
        return this.entityCollection;
    };
    /**
     * @return {?}
     */
    OptimizedEntityCollection.prototype.isFree = /**
     * @return {?}
     */
    function () {
        return this._collectionSize < 1 || this.entityCollection.values.length < this._collectionSize;
    };
    /**
     * @param {?} entity
     * @return {?}
     */
    OptimizedEntityCollection.prototype.add = /**
     * @param {?} entity
     * @return {?}
     */
    function (entity) {
        this.suspend();
        return this.entityCollection.add(entity);
    };
    /**
     * @param {?} entity
     * @return {?}
     */
    OptimizedEntityCollection.prototype.remove = /**
     * @param {?} entity
     * @return {?}
     */
    function (entity) {
        this.suspend();
        return this.entityCollection.remove(entity);
    };
    /**
     * @param {?} entity
     * @return {?}
     */
    OptimizedEntityCollection.prototype.removeNoSuspend = /**
     * @param {?} entity
     * @return {?}
     */
    function (entity) {
        this.entityCollection.remove(entity);
    };
    /**
     * @return {?}
     */
    OptimizedEntityCollection.prototype.removeAll = /**
     * @return {?}
     */
    function () {
        this.suspend();
        this.entityCollection.removeAll();
    };
    /**
     * @param {?} callback
     * @param {?=} once
     * @return {?}
     */
    OptimizedEntityCollection.prototype.onEventSuspension = /**
     * @param {?} callback
     * @param {?=} once
     * @return {?}
     */
    function (callback, once) {
        var _this = this;
        if (once === void 0) { once = false; }
        this._onEventSuspensionCallback = { callback: callback, once: once };
        return function () {
            _this._onEventSuspensionCallback = undefined;
        };
    };
    /**
     * @param {?} callback
     * @param {?=} once
     * @return {?}
     */
    OptimizedEntityCollection.prototype.onEventResume = /**
     * @param {?} callback
     * @param {?=} once
     * @return {?}
     */
    function (callback, once) {
        var _this = this;
        if (once === void 0) { once = false; }
        this._onEventResumeCallback = { callback: callback, once: once };
        if (!this._isSuspended) {
            this.triggerEventResume();
        }
        return function () {
            _this._onEventResumeCallback = undefined;
        };
    };
    /**
     * @return {?}
     */
    OptimizedEntityCollection.prototype.triggerEventSuspension = /**
     * @return {?}
     */
    function () {
        if (this._onEventSuspensionCallback !== undefined) {
            /** @type {?} */
            var callback = this._onEventSuspensionCallback.callback;
            if (this._onEventSuspensionCallback.once) {
                this._onEventSuspensionCallback = undefined;
            }
            callback();
        }
    };
    /**
     * @return {?}
     */
    OptimizedEntityCollection.prototype.triggerEventResume = /**
     * @return {?}
     */
    function () {
        if (this._onEventResumeCallback !== undefined) {
            /** @type {?} */
            var callback = this._onEventResumeCallback.callback;
            if (this._onEventResumeCallback.once) {
                this._onEventResumeCallback = undefined;
            }
            callback();
        }
    };
    /**
     * @return {?}
     */
    OptimizedEntityCollection.prototype.suspend = /**
     * @return {?}
     */
    function () {
        var _this = this;
        if (this._updateRate < 0) {
            return;
        }
        if (this._isHardSuspend) {
            return;
        }
        if (!this._isSuspended) {
            this._isSuspended = true;
            this.entityCollection.suspendEvents();
            this.triggerEventSuspension();
            this._suspensionTimeout = setTimeout(function () {
                _this.entityCollection.resumeEvents();
                _this.triggerEventResume();
                _this._isSuspended = false;
                _this._suspensionTimeout = undefined;
            }, this._updateRate);
        }
    };
    /**
     * @return {?}
     */
    OptimizedEntityCollection.prototype.hardSuspend = /**
     * @return {?}
     */
    function () {
        this.entityCollection.suspendEvents();
        this._isHardSuspend = true;
    };
    /**
     * @return {?}
     */
    OptimizedEntityCollection.prototype.hardResume = /**
     * @return {?}
     */
    function () {
        this.entityCollection.resumeEvents();
        this._isHardSuspend = false;
    };
    return OptimizedEntityCollection;
}());
export { OptimizedEntityCollection };
if (false) {
    /**
     * @type {?}
     * @private
     */
    OptimizedEntityCollection.prototype._updateRate;
    /**
     * @type {?}
     * @private
     */
    OptimizedEntityCollection.prototype._collectionSize;
    /**
     * @type {?}
     * @private
     */
    OptimizedEntityCollection.prototype._isSuspended;
    /**
     * @type {?}
     * @private
     */
    OptimizedEntityCollection.prototype._isHardSuspend;
    /**
     * @type {?}
     * @private
     */
    OptimizedEntityCollection.prototype._suspensionTimeout;
    /**
     * @type {?}
     * @private
     */
    OptimizedEntityCollection.prototype._onEventSuspensionCallback;
    /**
     * @type {?}
     * @private
     */
    OptimizedEntityCollection.prototype._onEventResumeCallback;
    /**
     * @type {?}
     * @private
     */
    OptimizedEntityCollection.prototype.entityCollection;
}
