// angular-cesium/esm5/lib/angular-cesium/components/ac-circle/ac-circle.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component } from '@angular/core';
import { EntityOnMapComponent } from '../../services/entity-on-map/entity-on-map.component';
import { EllipseDrawerService } from '../../services/drawers/ellipse-drawer/ellipse-drawer.service';
import { MapLayersService } from '../../services/map-layers/map-layers.service';
/**
 *  This is a circle implementation.
 *  The element must be a child of ac-map element.
 *  semiMajorAxis ans semiMinorAxis are replaced with radius property.
 *  All other properties of props are the same as the properties of Entity and EllipseGraphics:
 *  + https://cesiumjs.org/Cesium/Build/Documentation/Entity.html
 *  + https://cesiumjs.org/Cesium/Build/Documentation/EllipseGraphics.html
 *
 *  __Usage:__
 *  ```
 *  <ac-circle [props]="{
 *    position: position,
 *    radius:40000.0,
 *    granularity:0.03,
 *  }">
 *  </ac-circle>
 *  ```
 */
var AcCircleComponent = /** @class */ (function (_super) {
    tslib_1.__extends(AcCircleComponent, _super);
    function AcCircleComponent(ellipseDrawerService, mapLayers) {
        return _super.call(this, ellipseDrawerService, mapLayers) || this;
    }
    /**
     * @private
     * @return {?}
     */
    AcCircleComponent.prototype.updateEllipseProps = /**
     * @private
     * @return {?}
     */
    function () {
        this.props.semiMajorAxis = this.props.radius;
        this.props.semiMinorAxis = this.props.radius;
        this.props.rotation = 0.0;
    };
    /**
     * @return {?}
     */
    AcCircleComponent.prototype.drawOnMap = /**
     * @return {?}
     */
    function () {
        this.updateEllipseProps();
        _super.prototype.drawOnMap.call(this);
    };
    /**
     * @return {?}
     */
    AcCircleComponent.prototype.updateOnMap = /**
     * @return {?}
     */
    function () {
        this.updateEllipseProps();
        _super.prototype.updateOnMap.call(this);
    };
    AcCircleComponent.decorators = [
        { type: Component, args: [{
                    selector: 'ac-circle',
                    template: ''
                }] }
    ];
    /** @nocollapse */
    AcCircleComponent.ctorParameters = function () { return [
        { type: EllipseDrawerService },
        { type: MapLayersService }
    ]; };
    return AcCircleComponent;
}(EntityOnMapComponent));
export { AcCircleComponent };
