// angular-cesium/esm5/lib/angular-cesium/services/map-layers/map-layers.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { CesiumService } from '../cesium/cesium.service';
import { Injectable } from '@angular/core';
var MapLayersService = /** @class */ (function () {
    function MapLayersService(cesiumService) {
        this.cesiumService = cesiumService;
        this.layersDataSources = [];
    }
    /**
     * @param {?} dataSources
     * @param {?} zIndex
     * @return {?}
     */
    MapLayersService.prototype.registerLayerDataSources = /**
     * @param {?} dataSources
     * @param {?} zIndex
     * @return {?}
     */
    function (dataSources, zIndex) {
        var _this = this;
        dataSources.forEach(function (ds) {
            ds.zIndex = zIndex;
            _this.layersDataSources.push(ds);
        });
    };
    /**
     * @return {?}
     */
    MapLayersService.prototype.drawAllLayers = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.layersDataSources.sort(function (a, b) { return a.zIndex - b.zIndex; });
        this.layersDataSources.forEach(function (dataSource) {
            _this.cesiumService.getViewer().dataSources.add(dataSource);
        });
    };
    /**
     * @param {?} dataSources
     * @param {?} newZIndex
     * @return {?}
     */
    MapLayersService.prototype.updateAndRefresh = /**
     * @param {?} dataSources
     * @param {?} newZIndex
     * @return {?}
     */
    function (dataSources, newZIndex) {
        var _this = this;
        if (dataSources && dataSources.length) {
            dataSources.forEach(function (ds) {
                /** @type {?} */
                var index = _this.layersDataSources.indexOf(ds);
                if (index !== -1) {
                    _this.layersDataSources[index].zIndex = newZIndex;
                }
            });
            this.cesiumService.getViewer().dataSources.removeAll();
            this.drawAllLayers();
        }
    };
    /**
     * @param {?} dataSources
     * @return {?}
     */
    MapLayersService.prototype.removeDataSources = /**
     * @param {?} dataSources
     * @return {?}
     */
    function (dataSources) {
        var _this = this;
        dataSources.forEach(function (ds) {
            /** @type {?} */
            var index = _this.layersDataSources.indexOf(ds);
            if (index !== -1) {
                _this.layersDataSources.splice(index, 1);
                _this.cesiumService.getViewer().dataSources.remove(ds, true);
            }
        });
    };
    MapLayersService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    MapLayersService.ctorParameters = function () { return [
        { type: CesiumService }
    ]; };
    return MapLayersService;
}());
export { MapLayersService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    MapLayersService.prototype.layersDataSources;
    /**
     * @type {?}
     * @private
     */
    MapLayersService.prototype.cesiumService;
}
