// angular-cesium/esm5/lib/angular-cesium-widgets/models/edit-polyline.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { AcEntity } from '../../angular-cesium/models/ac-entity';
var EditPolyline = /** @class */ (function (_super) {
    tslib_1.__extends(EditPolyline, _super);
    function EditPolyline(entityId, startPosition, endPosition, polylineProps) {
        var _this = _super.call(this) || this;
        _this.editedEntityId = entityId;
        _this.id = _this.generateId();
        _this.positions = [startPosition, endPosition];
        _this._polylineProps = polylineProps;
        return _this;
    }
    Object.defineProperty(EditPolyline.prototype, "props", {
        get: /**
         * @return {?}
         */
        function () {
            return this._polylineProps;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._polylineProps = value;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    EditPolyline.prototype.getEditedEntityId = /**
     * @return {?}
     */
    function () {
        return this.editedEntityId;
    };
    /**
     * @return {?}
     */
    EditPolyline.prototype.getPositions = /**
     * @return {?}
     */
    function () {
        return this.positions;
    };
    /**
     * @return {?}
     */
    EditPolyline.prototype.validatePositions = /**
     * @return {?}
     */
    function () {
        return this.positions[0] !== undefined && this.positions[1] !== undefined;
    };
    /**
     * @return {?}
     */
    EditPolyline.prototype.getStartPosition = /**
     * @return {?}
     */
    function () {
        return this.positions[0];
    };
    /**
     * @return {?}
     */
    EditPolyline.prototype.getEndPosition = /**
     * @return {?}
     */
    function () {
        return this.positions[1];
    };
    /**
     * @param {?} position
     * @return {?}
     */
    EditPolyline.prototype.setStartPosition = /**
     * @param {?} position
     * @return {?}
     */
    function (position) {
        this.positions[0] = position;
    };
    /**
     * @param {?} position
     * @return {?}
     */
    EditPolyline.prototype.setEndPosition = /**
     * @param {?} position
     * @return {?}
     */
    function (position) {
        this.positions[1] = position;
    };
    /**
     * @return {?}
     */
    EditPolyline.prototype.getId = /**
     * @return {?}
     */
    function () {
        return this.id;
    };
    /**
     * @private
     * @return {?}
     */
    EditPolyline.prototype.generateId = /**
     * @private
     * @return {?}
     */
    function () {
        return 'edit-polyline-' + EditPolyline.counter++;
    };
    EditPolyline.counter = 0;
    return EditPolyline;
}(AcEntity));
export { EditPolyline };
if (false) {
    /** @type {?} */
    EditPolyline.counter;
    /**
     * @type {?}
     * @private
     */
    EditPolyline.prototype.editedEntityId;
    /**
     * @type {?}
     * @private
     */
    EditPolyline.prototype.id;
    /**
     * @type {?}
     * @private
     */
    EditPolyline.prototype.positions;
    /**
     * @type {?}
     * @private
     */
    EditPolyline.prototype._polylineProps;
}
