// angular-cesium/esm2015/lib/angular-cesium/directives/ac-html/ac-html.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectorRef, Directive, TemplateRef, ViewContainerRef } from '@angular/core';
import { LayerService } from '../../services/layer-service/layer-service.service';
import { AcHtmlManager } from '../../services/ac-html-manager/ac-html-manager.service';
export class AcHtmlContext {
    /**
     * @param {?} id
     * @param {?} context
     */
    constructor(id, context) {
        this.id = id;
        this.context = context;
    }
}
if (false) {
    /** @type {?} */
    AcHtmlContext.prototype.id;
    /** @type {?} */
    AcHtmlContext.prototype.context;
}
export class AcHtmlDirective {
    /**
     * @param {?} _templateRef
     * @param {?} _viewContainerRef
     * @param {?} _changeDetector
     * @param {?} _layerService
     * @param {?} _acHtmlManager
     */
    constructor(_templateRef, _viewContainerRef, _changeDetector, _layerService, _acHtmlManager) {
        this._templateRef = _templateRef;
        this._viewContainerRef = _viewContainerRef;
        this._changeDetector = _changeDetector;
        this._layerService = _layerService;
        this._acHtmlManager = _acHtmlManager;
        this._views = new Map();
    }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
    /**
     * @private
     * @param {?} id
     * @param {?} primitive
     * @param {?} entity
     * @return {?}
     */
    _handleView(id, primitive, entity) {
        if (!this._views.has(id) && primitive.show) {
            /** @type {?} */
            const context = new AcHtmlContext(id, { $implicit: entity });
            /** @type {?} */
            const viewRef = this._viewContainerRef.createEmbeddedView(this._templateRef, context);
            this._views.set(id, { viewRef, context });
            this._changeDetector.detectChanges();
        }
        else if (this._views.has(id) && !primitive.show) {
            this.remove(id, primitive);
        }
        else if (this._views.has(id) && primitive.show) {
            this._changeDetector.detectChanges();
        }
    }
    /**
     * @param {?} id
     * @param {?} primitive
     * @return {?}
     */
    addOrUpdate(id, primitive) {
        /** @type {?} */
        const context = this._layerService.context;
        /** @type {?} */
        const entity = context[this._layerService.getEntityName()];
        if (this._views.has(id)) {
            this._views.get(id).context.context.$implicit = entity;
        }
        this._acHtmlManager.addOrUpdate(id, { entity, primitive });
        this._handleView(id, primitive, entity);
    }
    /**
     * @param {?} id
     * @param {?} primitive
     * @return {?}
     */
    remove(id, primitive) {
        if (!this._views.has(id)) {
            return;
        }
        const { viewRef } = this._views.get(id);
        this._viewContainerRef.remove(this._viewContainerRef.indexOf(viewRef));
        this._views.delete(id);
        this._acHtmlManager.remove(id);
        primitive.element = null;
    }
}
AcHtmlDirective.decorators = [
    { type: Directive, args: [{
                selector: '[acHtml]',
            },] }
];
/** @nocollapse */
AcHtmlDirective.ctorParameters = () => [
    { type: TemplateRef },
    { type: ViewContainerRef },
    { type: ChangeDetectorRef },
    { type: LayerService },
    { type: AcHtmlManager }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    AcHtmlDirective.prototype._views;
    /**
     * @type {?}
     * @private
     */
    AcHtmlDirective.prototype._templateRef;
    /**
     * @type {?}
     * @private
     */
    AcHtmlDirective.prototype._viewContainerRef;
    /**
     * @type {?}
     * @private
     */
    AcHtmlDirective.prototype._changeDetector;
    /**
     * @type {?}
     * @private
     */
    AcHtmlDirective.prototype._layerService;
    /**
     * @type {?}
     * @private
     */
    AcHtmlDirective.prototype._acHtmlManager;
}
