// angular-cesium/esm2015/lib/angular-cesium/services/drawers/czml-drawer/czml-drawer.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { CesiumService } from '../../cesium/cesium.service';
import { BasicDrawerService } from '../basic-drawer/basic-drawer.service';
/**
 *  This drawer is responsible for drawing czml dataSources.
 */
export class CzmlDrawerService extends BasicDrawerService {
    /**
     * @param {?} cesiumService
     */
    constructor(cesiumService) {
        super();
        this.cesiumService = cesiumService;
    }
    /**
     * @param {?=} options
     * @return {?}
     */
    init(options) {
        /** @type {?} */
        const dataSources = [];
        this.czmlStream = new Cesium.CzmlDataSource('czml');
        dataSources.push(this.czmlStream);
        this.cesiumService.getViewer().dataSources.add(this.czmlStream);
        return dataSources;
    }
    // returns the packet, provided by the stream
    /**
     * @param {?} cesiumProps
     * @return {?}
     */
    add(cesiumProps) {
        this.czmlStream.process(cesiumProps.czmlPacket);
        return cesiumProps;
    }
    /**
     * @param {?} entity
     * @param {?} cesiumProps
     * @return {?}
     */
    update(entity, cesiumProps) {
        this.czmlStream.process(cesiumProps.czmlPacket);
    }
    /**
     * @param {?} entity
     * @return {?}
     */
    remove(entity) {
        this.czmlStream.entities.removeById(entity.acEntity.id);
    }
    /**
     * @return {?}
     */
    removeAll() {
        this.czmlStream.entities.removeAll();
    }
    /**
     * @param {?} showValue
     * @return {?}
     */
    setShow(showValue) {
        this.czmlStream.entities.show = showValue;
    }
}
CzmlDrawerService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
CzmlDrawerService.ctorParameters = () => [
    { type: CesiumService }
];
if (false) {
    /** @type {?} */
    CzmlDrawerService.prototype.czmlStream;
    /**
     * @type {?}
     * @private
     */
    CzmlDrawerService.prototype.cesiumService;
}
