// angular-cesium/esm2015/lib/angular-cesium-widgets/models/edit-arc.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { AcEntity } from '../../angular-cesium/models/ac-entity';
export class EditArc extends AcEntity {
    /**
     * @param {?} entityId
     * @param {?} center
     * @param {?} radius
     * @param {?} delta
     * @param {?} angle
     * @param {?} _arcProps
     */
    constructor(entityId, center, radius, delta, angle, _arcProps) {
        super();
        this._arcProps = _arcProps;
        this.id = this.generateId();
        this.editedEntityId = entityId;
        this._center = center;
        this._radius = radius;
        this._delta = delta;
        this._angle = angle;
    }
    /**
     * @return {?}
     */
    get props() {
        return this._arcProps;
    }
    /**
     * @param {?} props
     * @return {?}
     */
    set props(props) {
        this._arcProps = props;
    }
    /**
     * @return {?}
     */
    get angle() {
        return this._angle;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set angle(value) {
        this._angle = value;
    }
    /**
     * @return {?}
     */
    get delta() {
        return this._delta;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set delta(value) {
        this._delta = value;
    }
    /**
     * @return {?}
     */
    get radius() {
        return this._radius;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set radius(value) {
        this._radius = value;
    }
    /**
     * @return {?}
     */
    get center() {
        return this._center;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set center(value) {
        this._center = value;
    }
    /**
     * @param {?} center
     * @return {?}
     */
    updateCenter(center) {
        this._center.x = center.x;
        this._center.y = center.y;
        this._center.z = center.z;
    }
    /**
     * @return {?}
     */
    getId() {
        return this.id;
    }
    /**
     * @private
     * @return {?}
     */
    generateId() {
        return 'edit-arc-' + EditArc.counter++;
    }
}
EditArc.counter = 0;
if (false) {
    /** @type {?} */
    EditArc.counter;
    /**
     * @type {?}
     * @private
     */
    EditArc.prototype.id;
    /**
     * @type {?}
     * @private
     */
    EditArc.prototype.editedEntityId;
    /**
     * @type {?}
     * @private
     */
    EditArc.prototype._center;
    /**
     * @type {?}
     * @private
     */
    EditArc.prototype._radius;
    /**
     * @type {?}
     * @private
     */
    EditArc.prototype._delta;
    /**
     * @type {?}
     * @private
     */
    EditArc.prototype._angle;
    /**
     * @type {?}
     * @private
     */
    EditArc.prototype._arcProps;
}
