// angular-cesium/esm2015/lib/angular-cesium-widgets/models/edit-actions.enum.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @enum {number} */
const EditActions = {
    INIT: 0,
    MOUSE_MOVE: 1,
    ADD_POINT: 2,
    ADD_LAST_POINT: 3,
    CHANGE_TO_EDIT: 4,
    REMOVE_POINT: 5,
    DRAG_POINT: 6,
    DRAG_POINT_FINISH: 7,
    DRAG_SHAPE: 8,
    DRAG_SHAPE_FINISH: 9,
    DONE: 10,
    DISABLE: 11,
    ENABLE: 12,
    DISPOSE: 13,
    SET_EDIT_LABELS_RENDER_CALLBACK: 14,
    UPDATE_EDIT_LABELS: 15,
    SET_MANUALLY: 16,
    TRANSFORM: 17,
};
export { EditActions };
EditActions[EditActions.INIT] = 'INIT';
EditActions[EditActions.MOUSE_MOVE] = 'MOUSE_MOVE';
EditActions[EditActions.ADD_POINT] = 'ADD_POINT';
EditActions[EditActions.ADD_LAST_POINT] = 'ADD_LAST_POINT';
EditActions[EditActions.CHANGE_TO_EDIT] = 'CHANGE_TO_EDIT';
EditActions[EditActions.REMOVE_POINT] = 'REMOVE_POINT';
EditActions[EditActions.DRAG_POINT] = 'DRAG_POINT';
EditActions[EditActions.DRAG_POINT_FINISH] = 'DRAG_POINT_FINISH';
EditActions[EditActions.DRAG_SHAPE] = 'DRAG_SHAPE';
EditActions[EditActions.DRAG_SHAPE_FINISH] = 'DRAG_SHAPE_FINISH';
EditActions[EditActions.DONE] = 'DONE';
EditActions[EditActions.DISABLE] = 'DISABLE';
EditActions[EditActions.ENABLE] = 'ENABLE';
EditActions[EditActions.DISPOSE] = 'DISPOSE';
EditActions[EditActions.SET_EDIT_LABELS_RENDER_CALLBACK] = 'SET_EDIT_LABELS_RENDER_CALLBACK';
EditActions[EditActions.UPDATE_EDIT_LABELS] = 'UPDATE_EDIT_LABELS';
EditActions[EditActions.SET_MANUALLY] = 'SET_MANUALLY';
EditActions[EditActions.TRANSFORM] = 'TRANSFORM';
