// angular-cesium/esm2015/lib/angular-cesium/services/drawers/primitives-drawer/primitives-drawer.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BasicDrawerService } from '../basic-drawer/basic-drawer.service';
/**
 *  General primitives drawer responsible of drawing Cesium primitives.
 *  Drawers the handle Cesium primitives extend it.
 * @abstract
 */
export class PrimitivesDrawerService extends BasicDrawerService {
    /**
     * @param {?} drawerType
     * @param {?} cesiumService
     */
    constructor(drawerType, cesiumService) {
        super();
        this.drawerType = drawerType;
        this.cesiumService = cesiumService;
        this._show = true;
    }
    /**
     * @return {?}
     */
    init() {
        this._cesiumCollection = new this.drawerType();
        this._primitiveCollectionWrap = new Cesium.PrimitiveCollection();
        this._primitiveCollectionWrap.add(this._cesiumCollection);
        this.cesiumService.getScene().primitives.add(this._primitiveCollectionWrap);
    }
    /**
     * @param {?} cesiumProps
     * @param {...?} args
     * @return {?}
     */
    add(cesiumProps, ...args) {
        return this._cesiumCollection.add(cesiumProps);
    }
    /**
     * @param {?} entity
     * @param {?} cesiumProps
     * @param {...?} args
     * @return {?}
     */
    update(entity, cesiumProps, ...args) {
        if (this._propsAssigner) {
            this._propsAssigner(entity, cesiumProps);
        }
        else {
            Object.assign(entity, cesiumProps);
        }
    }
    /**
     * @param {?} entity
     * @return {?}
     */
    remove(entity) {
        this._cesiumCollection.remove(entity);
    }
    /**
     * @return {?}
     */
    removeAll() {
        this._cesiumCollection.removeAll();
    }
    /**
     * @param {?} showValue
     * @return {?}
     */
    setShow(showValue) {
        this._show = showValue;
        this._primitiveCollectionWrap.show = showValue;
    }
    /**
     * @return {?}
     */
    getShow() {
        return this._show;
    }
}
if (false) {
    /**
     * @type {?}
     * @private
     */
    PrimitivesDrawerService.prototype._show;
    /**
     * @type {?}
     * @private
     */
    PrimitivesDrawerService.prototype._primitiveCollectionWrap;
    /**
     * @type {?}
     * @protected
     */
    PrimitivesDrawerService.prototype._cesiumCollection;
    /**
     * @type {?}
     * @protected
     */
    PrimitivesDrawerService.prototype._propsAssigner;
    /**
     * @type {?}
     * @private
     */
    PrimitivesDrawerService.prototype.drawerType;
    /**
     * @type {?}
     * @private
     */
    PrimitivesDrawerService.prototype.cesiumService;
}
