// angular-cesium/esm5/lib/angular-cesium/services/layer-service/layer-service.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { EventEmitter, Injectable } from '@angular/core';
var LayerService = /** @class */ (function () {
    function LayerService() {
        this._cache = true;
        this.descriptions = [];
        this.layerUpdate = new EventEmitter();
    }
    Object.defineProperty(LayerService.prototype, "cache", {
        get: /**
         * @return {?}
         */
        function () {
            return this._cache;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._cache = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(LayerService.prototype, "zIndex", {
        get: /**
         * @return {?}
         */
        function () {
            return this._zIndex;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            if (value !== this._zIndex) {
                this.layerUpdate.emit();
            }
            this._zIndex = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(LayerService.prototype, "show", {
        get: /**
         * @return {?}
         */
        function () {
            return this._show;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            if (value !== this._show) {
                this.layerUpdate.emit();
            }
            this._show = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(LayerService.prototype, "options", {
        get: /**
         * @return {?}
         */
        function () {
            return this._options;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._options = value;
            this.layerUpdate.emit();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(LayerService.prototype, "context", {
        get: /**
         * @return {?}
         */
        function () {
            return this._context;
        },
        set: /**
         * @param {?} context
         * @return {?}
         */
        function (context) {
            this._context = context;
            this.layerUpdate.emit();
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} name
     * @return {?}
     */
    LayerService.prototype.setEntityName = /**
     * @param {?} name
     * @return {?}
     */
    function (name) {
        this._entityName = name;
    };
    /**
     * @return {?}
     */
    LayerService.prototype.getEntityName = /**
     * @return {?}
     */
    function () {
        return this._entityName;
    };
    /**
     * @param {?} descriptionComponent
     * @return {?}
     */
    LayerService.prototype.registerDescription = /**
     * @param {?} descriptionComponent
     * @return {?}
     */
    function (descriptionComponent) {
        if (this.descriptions.indexOf(descriptionComponent) < 0) {
            this.descriptions.push(descriptionComponent);
        }
    };
    /**
     * @param {?} descriptionComponent
     * @return {?}
     */
    LayerService.prototype.unregisterDescription = /**
     * @param {?} descriptionComponent
     * @return {?}
     */
    function (descriptionComponent) {
        /** @type {?} */
        var index = this.descriptions.indexOf(descriptionComponent);
        if (index > -1) {
            this.descriptions.splice(index, 1);
        }
    };
    /**
     * @return {?}
     */
    LayerService.prototype.getDescriptions = /**
     * @return {?}
     */
    function () {
        return this.descriptions;
    };
    /**
     * @return {?}
     */
    LayerService.prototype.layerUpdates = /**
     * @return {?}
     */
    function () {
        return this.layerUpdate;
    };
    LayerService.decorators = [
        { type: Injectable }
    ];
    return LayerService;
}());
export { LayerService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    LayerService.prototype._context;
    /**
     * @type {?}
     * @private
     */
    LayerService.prototype._options;
    /**
     * @type {?}
     * @private
     */
    LayerService.prototype._show;
    /**
     * @type {?}
     * @private
     */
    LayerService.prototype._zIndex;
    /**
     * @type {?}
     * @private
     */
    LayerService.prototype._entityName;
    /**
     * @type {?}
     * @private
     */
    LayerService.prototype._cache;
    /**
     * @type {?}
     * @private
     */
    LayerService.prototype.descriptions;
    /**
     * @type {?}
     * @private
     */
    LayerService.prototype.layerUpdate;
}
