// angular-cesium/esm5/lib/angular-cesium/components/ac-arc/ac-arc.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, Input } from '@angular/core';
import { EntityOnMapComponent } from '../../services/entity-on-map/entity-on-map.component';
import { ArcDrawerService } from '../../services/drawers/arc-drawer/arc-drawer.service';
import { MapLayersService } from '../../services/map-layers/map-layers.service';
/**
 *  This is an implementation of an arc.
 *  The element must be a child of ac-map element.
 *  An arc is not natively implemented in cesium.
 *
 *  __Usage :__
 *  ```
 *    <ac-arc-desc geometryProps="{
 *          center: arc.center,
 *          angle: arc.angle,
 *          delta: arc.delta,
 *          radius: arc.radius
 *       }"
 *       instanceProps="{
 *          attributes: arc.attributes
 *       }"
 *       primitiveProps="{
 *          appearance: arc.appearance
 *       }">
 *    </ac-arc-desc>
 *    ```
 */
var AcArcComponent = /** @class */ (function (_super) {
    tslib_1.__extends(AcArcComponent, _super);
    function AcArcComponent(arcDrawer, mapLayers) {
        return _super.call(this, arcDrawer, mapLayers) || this;
    }
    /**
     * @return {?}
     */
    AcArcComponent.prototype.updateOnMap = /**
     * @return {?}
     */
    function () {
        if (this.selfPrimitiveIsDraw) {
            this.removeFromMap();
            this.drawOnMap();
        }
    };
    /**
     * @return {?}
     */
    AcArcComponent.prototype.drawOnMap = /**
     * @return {?}
     */
    function () {
        this.selfPrimitiveIsDraw = true;
        return this.selfPrimitive = this._drawer.add(this.geometryProps, this.instanceProps, this.primitiveProps);
    };
    /**
     * @param {?} changes
     * @return {?}
     */
    AcArcComponent.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        /** @type {?} */
        var geometryProps = changes['geometryProps'];
        /** @type {?} */
        var instanceProps = changes['instanceProps'];
        /** @type {?} */
        var primitiveProps = changes['primitiveProps'];
        if (geometryProps.currentValue !== geometryProps.previousValue ||
            instanceProps.currentValue !== instanceProps.previousValue ||
            primitiveProps.currentValue !== primitiveProps.previousValue) {
            this.updateOnMap();
        }
    };
    AcArcComponent.decorators = [
        { type: Component, args: [{
                    selector: 'ac-arc',
                    template: ''
                }] }
    ];
    /** @nocollapse */
    AcArcComponent.ctorParameters = function () { return [
        { type: ArcDrawerService },
        { type: MapLayersService }
    ]; };
    AcArcComponent.propDecorators = {
        geometryProps: [{ type: Input }],
        instanceProps: [{ type: Input }],
        primitiveProps: [{ type: Input }]
    };
    return AcArcComponent;
}(EntityOnMapComponent));
export { AcArcComponent };
if (false) {
    /** @type {?} */
    AcArcComponent.prototype.geometryProps;
    /** @type {?} */
    AcArcComponent.prototype.instanceProps;
    /** @type {?} */
    AcArcComponent.prototype.primitiveProps;
}
