// angular-cesium/esm2015/lib/angular-cesium/services/maps-manager/maps-manager.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
/**
 *  The service manages `ac-map` instances. `ac-map` register itself to this service.
 *  This allows retrieval of maps provided services outside of `ac-map` scope.
 */
export class MapsManagerService {
    constructor() {
        this.defaultIdCounter = 0;
        this._Maps = new Map();
        this.eventRemoveCallbacks = [];
    }
    /**
     * @param {?=} id
     * @return {?}
     */
    getMap(id) {
        if (!id) {
            return this.firstMap;
        }
        return this._Maps.get(id);
    }
    /**
     * @param {?} id
     * @param {?} acMap
     * @return {?}
     */
    _registerMap(id, acMap) {
        if (!this.firstMap) {
            this.firstMap = acMap;
        }
        /** @type {?} */
        const mapId = id ? id : this.generateDefaultId();
        if (this._Maps.has(mapId)) {
            throw new Error(`Map with id: ${mapId} already exist`);
        }
        this._Maps.set(mapId, acMap);
        return mapId;
    }
    /**
     * @param {?} id
     * @return {?}
     */
    _removeMapById(id) {
        return this._Maps.delete(id);
    }
    /**
     * @private
     * @return {?}
     */
    generateDefaultId() {
        this.defaultIdCounter++;
        return 'default-map-id-' + this.defaultIdCounter;
    }
    /**
     * Binds multiple 2D map's cameras together.
     * @param {?} mapsConfiguration - binding options.
     * mapId - the id of the maps to bind.
     * sensitivity - the amount the camera position should change in order to sync other maps.
     * bindZoom - should bind zoom level
     * @return {?}
     */
    sync2DMapsCameras(mapsConfiguration) {
        /** @type {?} */
        const DEFAULT_SENSITIVITY = 0.01;
        this.unsyncMapsCameras();
        /** @type {?} */
        const maps = mapsConfiguration.map(config => {
            /** @type {?} */
            const map = this.getMap(config.id);
            if (!map) {
                throw new Error(`Couldn't find map with id: ${config.id}`);
            }
            return { map, options: { sensitivity: config.sensitivity, bindZoom: config.bindZoom } };
        });
        maps.forEach(masterMapConfig => {
            /** @type {?} */
            const masterMap = masterMapConfig.map;
            /** @type {?} */
            const options = masterMapConfig.options;
            /** @type {?} */
            const masterCamera = masterMap.getCameraService().getCamera();
            /** @type {?} */
            const masterCameraCartographic = masterCamera.positionCartographic;
            masterCamera.percentageChanged = options.sensitivity || DEFAULT_SENSITIVITY;
            /** @type {?} */
            const removeCallback = masterCamera.changed.addEventListener(() => {
                maps.forEach(slaveMapConfig => {
                    /** @type {?} */
                    const slaveMap = slaveMapConfig.map;
                    /** @type {?} */
                    const slaveMapOptions = slaveMapConfig.options;
                    if (slaveMap === masterMap) {
                        return;
                    }
                    /** @type {?} */
                    const slaveCamera = slaveMap.getCameraService().getCamera();
                    /** @type {?} */
                    const slaveCameraCartographic = slaveCamera.positionCartographic;
                    /** @type {?} */
                    const position = Cesium.Ellipsoid.WGS84.cartographicToCartesian({
                        longitude: masterCameraCartographic.longitude,
                        latitude: masterCameraCartographic.latitude,
                        height: slaveMapOptions.bindZoom ? masterCameraCartographic.height : slaveCameraCartographic.height,
                    });
                    if (slaveMap.getCesiumViewer().scene.mode !== Cesium.SceneMode.MORPHING) {
                        slaveCamera.setView({
                            destination: position,
                            orientation: {
                                heading: slaveCamera.heading,
                                pitch: slaveCamera.pitch,
                            },
                        });
                    }
                });
            });
            this.eventRemoveCallbacks.push(removeCallback);
        });
    }
    /**
     * Unsyncs maps cameras
     * @return {?}
     */
    unsyncMapsCameras() {
        this.eventRemoveCallbacks.forEach(removeCallback => removeCallback());
        this.eventRemoveCallbacks = [];
    }
}
MapsManagerService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
MapsManagerService.ctorParameters = () => [];
if (false) {
    /**
     * @type {?}
     * @private
     */
    MapsManagerService.prototype.defaultIdCounter;
    /**
     * @type {?}
     * @private
     */
    MapsManagerService.prototype._Maps;
    /**
     * @type {?}
     * @private
     */
    MapsManagerService.prototype.firstMap;
    /**
     * @type {?}
     * @private
     */
    MapsManagerService.prototype.eventRemoveCallbacks;
}
