// angular-cesium/esm5/lib/angular-cesium-widgets/services/zoom-to-rectangle.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, Optional } from '@angular/core';
import { MapsManagerService } from '../../angular-cesium/services/maps-manager/maps-manager.service';
import { CameraService } from '../../angular-cesium/services/camera/camera.service';
import { CesiumService } from '../../angular-cesium/services/cesium/cesium.service';
/**
 * @record
 */
function ZoomData() { }
if (false) {
    /** @type {?} */
    ZoomData.prototype.container;
    /** @type {?|undefined} */
    ZoomData.prototype.borderElement;
    /** @type {?|undefined} */
    ZoomData.prototype.resetOnEscapePressFunc;
}
/** @enum {number} */
var MouseButtons = {
    LEFT: 0,
    MIDDLE: 1,
    RIGHT: 2,
};
export { MouseButtons };
MouseButtons[MouseButtons.LEFT] = 'LEFT';
MouseButtons[MouseButtons.MIDDLE] = 'MIDDLE';
MouseButtons[MouseButtons.RIGHT] = 'RIGHT';
var ZoomToRectangleService = /** @class */ (function () {
    function ZoomToRectangleService(mapsManager, cameraService, cesiumService) {
        this.mapsManager = mapsManager;
        this.mapsZoomElements = new Map();
        this.defaultOptions = {
            animationDurationInSeconds: 0.5,
            resetKeyCode: 27,
            borderStyle: '2px solid rgba(0,0,0,0.5)',
            backgroundColor: 'rgba(0,0,0,0.2)',
            autoDisableOnZoom: true,
            threshold: 9,
            keepRotation: true,
            mouseButton: MouseButtons.LEFT,
        };
    }
    /**
     * @param {?} cesiumService
     * @param {?} cameraService
     * @return {?}
     */
    ZoomToRectangleService.prototype.init = /**
     * @param {?} cesiumService
     * @param {?} cameraService
     * @return {?}
     */
    function (cesiumService, cameraService) {
        this.cameraService = cameraService;
        this.cesiumService = cesiumService;
    };
    /**
     * @param {?=} options
     * @param {?=} mapId
     * @return {?}
     */
    ZoomToRectangleService.prototype.activate = /**
     * @param {?=} options
     * @param {?=} mapId
     * @return {?}
     */
    function (options, mapId) {
        var _this = this;
        if (options === void 0) { options = {}; }
        if ((!this.cameraService || !this.cesiumService) && !mapId) {
            throw new Error("The function must receive a mapId if the service wasn't initialized");
        }
        /** @type {?} */
        var finalOptions = Object.assign({}, this.defaultOptions, options);
        /** @type {?} */
        var cameraService = this.cameraService;
        /** @type {?} */
        var mapContainer;
        /** @type {?} */
        var map;
        if (this.cesiumService) {
            mapContainer = this.cesiumService.getViewer().container;
            map = this.cesiumService.getMap();
        }
        if (mapId) {
            map = this.mapsManager.getMap(mapId);
            if (!map) {
                throw new Error("Map not found with id: " + mapId);
            }
            cameraService = map.getCameraService();
            mapContainer = map.getCesiumViewer().container;
        }
        if (!cameraService || !mapContainer) {
            throw new Error("The function must receive a mapId if the service wasn't initialized");
        }
        this.disable(mapId);
        /** @type {?} */
        var container = document.createElement('div');
        mapContainer.style.position = 'relative';
        container.style.position = 'absolute';
        container.style.width = '100%';
        container.style.height = '100%';
        container.style.top = '0';
        container.style.left = '0';
        mapContainer.appendChild(container);
        /** @type {?} */
        var mapZoomData = { container: container };
        this.mapsZoomElements.set(mapId || this.cesiumService.getMap().getId(), mapZoomData);
        /** @type {?} */
        var mouse = {
            endX: 0,
            endY: 0,
            startX: 0,
            startY: 0,
        };
        /** @type {?} */
        var borderElement;
        container.onmousedown = function (e) {
            if (e.button !== finalOptions.mouseButton) {
                return;
            }
            if (!borderElement) {
                if (options && options.onStart) {
                    options.onStart(map);
                }
                /** @type {?} */
                var rect = ((/** @type {?} */ (e.currentTarget))).getBoundingClientRect();
                /** @type {?} */
                var offsetX = e.clientX - rect.left;
                /** @type {?} */
                var offsetY = e.clientY - rect.top;
                mouse.startX = offsetX;
                mouse.startY = offsetY;
                borderElement = document.createElement('div');
                borderElement.className = 'zoom-to-rectangle-border';
                borderElement.style.position = 'absolute';
                borderElement.style.border = finalOptions.borderStyle;
                borderElement.style.backgroundColor = finalOptions.backgroundColor;
                borderElement.style.left = mouse.startX + 'px';
                borderElement.style.top = mouse.startY + 'px';
                container.appendChild(borderElement);
                mapZoomData.borderElement = borderElement;
            }
        };
        container.onmouseup = function (e) {
            if (borderElement) {
                /** @type {?} */
                var zoomApplied = void 0;
                if (mouse && Math.abs(mouse.endX - mouse.startX) * Math.abs(mouse.endY - mouse.startY) > finalOptions.threshold) {
                    zoomApplied = _this.zoomCameraToRectangle(cameraService, mouse, finalOptions.animationDurationInSeconds, finalOptions);
                }
                borderElement.remove();
                borderElement = undefined;
                mapZoomData.borderElement = undefined;
                mouse = {
                    endX: 0,
                    endY: 0,
                    startX: 0,
                    startY: 0,
                };
                if (!!finalOptions.onComplete) {
                    finalOptions.onComplete(map);
                }
                if (finalOptions.autoDisableOnZoom && zoomApplied) {
                    _this.disable(mapId);
                }
            }
        };
        container.onmousemove = function (e) {
            if (borderElement) {
                /** @type {?} */
                var rect = ((/** @type {?} */ (e.currentTarget))).getBoundingClientRect();
                /** @type {?} */
                var offsetX = e.clientX - rect.left;
                /** @type {?} */
                var offsetY = e.clientY - rect.top;
                mouse.endX = offsetX;
                mouse.endY = offsetY;
                borderElement.style.width = Math.abs(mouse.endX - mouse.startX) + 'px';
                borderElement.style.height = Math.abs(mouse.endY - mouse.startY) + 'px';
                borderElement.style.left = Math.min(mouse.startX, mouse.endX) + 'px';
                borderElement.style.top = Math.min(mouse.startY, mouse.endY) + 'px';
            }
        };
        /** @type {?} */
        var resetOnEscapePress = function (e) {
            if (e.keyCode === finalOptions.resetKeyCode && borderElement) {
                borderElement.remove();
                borderElement = undefined;
                mapZoomData.borderElement = undefined;
                mouse = {
                    endX: 0,
                    endY: 0,
                    startX: 0,
                    startY: 0,
                };
            }
        };
        document.addEventListener('keydown', resetOnEscapePress);
        mapZoomData.resetOnEscapePressFunc = resetOnEscapePress;
    };
    /**
     * @param {?=} mapId
     * @return {?}
     */
    ZoomToRectangleService.prototype.disable = /**
     * @param {?=} mapId
     * @return {?}
     */
    function (mapId) {
        if (!this.cesiumService && !mapId) {
            throw new Error('If the service was not initialized with CesiumService, mapId must be provided');
        }
        /** @type {?} */
        var data = this.mapsZoomElements.get(mapId || this.cesiumService.getMap().getId());
        if (data) {
            data.container.remove();
            if (data.borderElement) {
                data.borderElement.remove();
            }
            if (data.resetOnEscapePressFunc) {
                document.removeEventListener('keydown', data.resetOnEscapePressFunc);
            }
        }
        this.mapsZoomElements.delete(mapId);
    };
    /**
     * @private
     * @param {?} cameraService
     * @param {?} positions
     * @param {?} animationDuration
     * @param {?} options
     * @return {?}
     */
    ZoomToRectangleService.prototype.zoomCameraToRectangle = /**
     * @private
     * @param {?} cameraService
     * @param {?} positions
     * @param {?} animationDuration
     * @param {?} options
     * @return {?}
     */
    function (cameraService, positions, animationDuration, options) {
        /** @type {?} */
        var camera = cameraService.getCamera();
        /** @type {?} */
        var cartesian1 = camera.pickEllipsoid({ x: positions.startX, y: positions.startY });
        /** @type {?} */
        var cartesian2 = camera.pickEllipsoid({ x: positions.endX, y: positions.endY });
        if (!cartesian1 || !cartesian2) {
            return false;
        }
        /** @type {?} */
        var cartographic1 = Cesium.Cartographic.fromCartesian(cartesian1);
        /** @type {?} */
        var cartographic2 = Cesium.Cartographic.fromCartesian(cartesian2);
        cameraService.cameraFlyTo({
            destination: new Cesium.Rectangle(Math.min(cartographic1.longitude, cartographic2.longitude), Math.min(cartographic1.latitude, cartographic2.latitude), Math.max(cartographic1.longitude, cartographic2.longitude), Math.max(cartographic1.latitude, cartographic2.latitude)),
            orientation: options.keepRotation ? { heading: camera.heading } : undefined,
            duration: animationDuration,
        });
        return true;
    };
    ZoomToRectangleService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    ZoomToRectangleService.ctorParameters = function () { return [
        { type: MapsManagerService },
        { type: CameraService, decorators: [{ type: Optional }] },
        { type: CesiumService, decorators: [{ type: Optional }] }
    ]; };
    return ZoomToRectangleService;
}());
export { ZoomToRectangleService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    ZoomToRectangleService.prototype.cameraService;
    /**
     * @type {?}
     * @private
     */
    ZoomToRectangleService.prototype.cesiumService;
    /**
     * @type {?}
     * @private
     */
    ZoomToRectangleService.prototype.mapsZoomElements;
    /**
     * @type {?}
     * @private
     */
    ZoomToRectangleService.prototype.defaultOptions;
    /**
     * @type {?}
     * @private
     */
    ZoomToRectangleService.prototype.mapsManager;
}
