// angular-cesium/esm2015/lib/angular-cesium/services/layer-service/layer-service.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { EventEmitter, Injectable } from '@angular/core';
export class LayerService {
    constructor() {
        this._cache = true;
        this.descriptions = [];
        this.layerUpdate = new EventEmitter();
    }
    /**
     * @return {?}
     */
    get cache() {
        return this._cache;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set cache(value) {
        this._cache = value;
    }
    /**
     * @return {?}
     */
    get zIndex() {
        return this._zIndex;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set zIndex(value) {
        if (value !== this._zIndex) {
            this.layerUpdate.emit();
        }
        this._zIndex = value;
    }
    /**
     * @return {?}
     */
    get show() {
        return this._show;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set show(value) {
        if (value !== this._show) {
            this.layerUpdate.emit();
        }
        this._show = value;
    }
    /**
     * @return {?}
     */
    get options() {
        return this._options;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set options(value) {
        this._options = value;
        this.layerUpdate.emit();
    }
    /**
     * @return {?}
     */
    get context() {
        return this._context;
    }
    /**
     * @param {?} context
     * @return {?}
     */
    set context(context) {
        this._context = context;
        this.layerUpdate.emit();
    }
    /**
     * @param {?} name
     * @return {?}
     */
    setEntityName(name) {
        this._entityName = name;
    }
    /**
     * @return {?}
     */
    getEntityName() {
        return this._entityName;
    }
    /**
     * @param {?} descriptionComponent
     * @return {?}
     */
    registerDescription(descriptionComponent) {
        if (this.descriptions.indexOf(descriptionComponent) < 0) {
            this.descriptions.push(descriptionComponent);
        }
    }
    /**
     * @param {?} descriptionComponent
     * @return {?}
     */
    unregisterDescription(descriptionComponent) {
        /** @type {?} */
        const index = this.descriptions.indexOf(descriptionComponent);
        if (index > -1) {
            this.descriptions.splice(index, 1);
        }
    }
    /**
     * @return {?}
     */
    getDescriptions() {
        return this.descriptions;
    }
    /**
     * @return {?}
     */
    layerUpdates() {
        return this.layerUpdate;
    }
}
LayerService.decorators = [
    { type: Injectable }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    LayerService.prototype._context;
    /**
     * @type {?}
     * @private
     */
    LayerService.prototype._options;
    /**
     * @type {?}
     * @private
     */
    LayerService.prototype._show;
    /**
     * @type {?}
     * @private
     */
    LayerService.prototype._zIndex;
    /**
     * @type {?}
     * @private
     */
    LayerService.prototype._entityName;
    /**
     * @type {?}
     * @private
     */
    LayerService.prototype._cache;
    /**
     * @type {?}
     * @private
     */
    LayerService.prototype.descriptions;
    /**
     * @type {?}
     * @private
     */
    LayerService.prototype.layerUpdate;
}
