// angular-cesium/esm5/lib/angular-cesium/services/viewer-configuration/viewer-configuration.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
/**
 * Service for setting cesium viewer map options.
 * defaulty angular-cesium doesnt provide this service and viewer is created with default options.
 * In order set specific options you must set this service as provider in your component and
 * set the wanted options.
 * ```typescript
 * constructor(viewerConf :ViewerConfiguration ) {
 *   viewerConf.viewerOptions = { timeline: false };
 * }
 * ```
 * notice this configuration will be for all <ac-maps> in your component.
 */
var ViewerConfiguration = /** @class */ (function () {
    function ViewerConfiguration() {
        this.nextViewerOptionsIndex = 0;
        this.nextViewerModifierIndex = 0;
    }
    Object.defineProperty(ViewerConfiguration.prototype, "viewerOptions", {
        get: /**
         * @return {?}
         */
        function () {
            return this._viewerOptions;
        },
        /**
         * Can be used to set initial map viewer options.
         * If there is more than one map you can give the function an array of options.
         * The map initialized first will be set with the first option object in the options array and so on.
         */
        set: /**
         * Can be used to set initial map viewer options.
         * If there is more than one map you can give the function an array of options.
         * The map initialized first will be set with the first option object in the options array and so on.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._viewerOptions = value;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    ViewerConfiguration.prototype.getNextViewerOptions = /**
     * @return {?}
     */
    function () {
        if (this._viewerOptions instanceof Array) {
            return this._viewerOptions[this.nextViewerOptionsIndex++];
        }
        else {
            return this._viewerOptions;
        }
    };
    Object.defineProperty(ViewerConfiguration.prototype, "viewerModifier", {
        get: /**
         * @return {?}
         */
        function () {
            return this._viewerModifier;
        },
        /**
         * Can be used to set map viewer options after the map has been initialized.
         * If there is more than one map you can give the function an array of functions.
         * The map initialized first will be set with the first option object in the options array and so on.
         */
        set: /**
         * Can be used to set map viewer options after the map has been initialized.
         * If there is more than one map you can give the function an array of functions.
         * The map initialized first will be set with the first option object in the options array and so on.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._viewerModifier = value;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    ViewerConfiguration.prototype.getNextViewerModifier = /**
     * @return {?}
     */
    function () {
        if (this._viewerModifier instanceof Array) {
            return this._viewerModifier[this.nextViewerModifierIndex++];
        }
        else {
            return this._viewerModifier;
        }
    };
    ViewerConfiguration.decorators = [
        { type: Injectable }
    ];
    return ViewerConfiguration;
}());
export { ViewerConfiguration };
if (false) {
    /**
     * cesium viewer options According to [Viewer]{\@link https://cesiumjs.org/Cesium/Build/Documentation/Viewer.html?classFilter=vie}
     * @type {?}
     * @private
     */
    ViewerConfiguration.prototype._viewerOptions;
    /**
     * @type {?}
     * @private
     */
    ViewerConfiguration.prototype._viewerModifier;
    /**
     * @type {?}
     * @private
     */
    ViewerConfiguration.prototype.nextViewerOptionsIndex;
    /**
     * @type {?}
     * @private
     */
    ViewerConfiguration.prototype.nextViewerModifierIndex;
    /** @type {?} */
    ViewerConfiguration.prototype.Ï;
}
