// angular-cesium/esm5/lib/angular-cesium/services/entity-on-map/entity-on-map.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Input } from '@angular/core';
/**
 *  Extend this class to create drawing on map components.
 */
var EntityOnMapComponent = /** @class */ (function () {
    function EntityOnMapComponent(_drawer, mapLayers) {
        this._drawer = _drawer;
        this.mapLayers = mapLayers;
    }
    /**
     * @return {?}
     */
    EntityOnMapComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.selfPrimitiveIsDraw = false;
        /** @type {?} */
        var dataSources = this._drawer.init();
        if (dataSources) {
            this.dataSources = dataSources;
            // this.mapLayers.registerLayerDataSources(dataSources, 0);
        }
        this.drawOnMap();
    };
    /**
     * @param {?} changes
     * @return {?}
     */
    EntityOnMapComponent.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        /** @type {?} */
        var props = changes['props'];
        if (props.currentValue !== props.previousValue) {
            this.updateOnMap();
        }
    };
    /**
     * @return {?}
     */
    EntityOnMapComponent.prototype.drawOnMap = /**
     * @return {?}
     */
    function () {
        this.selfPrimitiveIsDraw = true;
        return this.selfPrimitive = this._drawer.add(this.props);
    };
    /**
     * @return {?}
     */
    EntityOnMapComponent.prototype.removeFromMap = /**
     * @return {?}
     */
    function () {
        this.selfPrimitiveIsDraw = false;
        return this._drawer.remove(this.selfPrimitive);
    };
    /**
     * @return {?}
     */
    EntityOnMapComponent.prototype.updateOnMap = /**
     * @return {?}
     */
    function () {
        if (this.selfPrimitiveIsDraw) {
            return this._drawer.update(this.selfPrimitive, this.props);
        }
    };
    /**
     * @return {?}
     */
    EntityOnMapComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.mapLayers.removeDataSources(this.dataSources);
        this.removeFromMap();
    };
    EntityOnMapComponent.propDecorators = {
        props: [{ type: Input }]
    };
    return EntityOnMapComponent;
}());
export { EntityOnMapComponent };
if (false) {
    /** @type {?} */
    EntityOnMapComponent.prototype.props;
    /**
     * @type {?}
     * @protected
     */
    EntityOnMapComponent.prototype.selfPrimitive;
    /**
     * @type {?}
     * @protected
     */
    EntityOnMapComponent.prototype.selfPrimitiveIsDraw;
    /**
     * @type {?}
     * @protected
     */
    EntityOnMapComponent.prototype.dataSources;
    /**
     * @type {?}
     * @protected
     */
    EntityOnMapComponent.prototype._drawer;
    /**
     * @type {?}
     * @private
     */
    EntityOnMapComponent.prototype.mapLayers;
}
