// angular-cesium/esm5/lib/angular-cesium/directives/ac-html/ac-html.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectorRef, Directive, TemplateRef, ViewContainerRef } from '@angular/core';
import { LayerService } from '../../services/layer-service/layer-service.service';
import { AcHtmlManager } from '../../services/ac-html-manager/ac-html-manager.service';
var AcHtmlContext = /** @class */ (function () {
    function AcHtmlContext(id, context) {
        this.id = id;
        this.context = context;
    }
    return AcHtmlContext;
}());
export { AcHtmlContext };
if (false) {
    /** @type {?} */
    AcHtmlContext.prototype.id;
    /** @type {?} */
    AcHtmlContext.prototype.context;
}
var AcHtmlDirective = /** @class */ (function () {
    function AcHtmlDirective(_templateRef, _viewContainerRef, _changeDetector, _layerService, _acHtmlManager) {
        this._templateRef = _templateRef;
        this._viewContainerRef = _viewContainerRef;
        this._changeDetector = _changeDetector;
        this._layerService = _layerService;
        this._acHtmlManager = _acHtmlManager;
        this._views = new Map();
    }
    /**
     * @return {?}
     */
    AcHtmlDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @private
     * @param {?} id
     * @param {?} primitive
     * @param {?} entity
     * @return {?}
     */
    AcHtmlDirective.prototype._handleView = /**
     * @private
     * @param {?} id
     * @param {?} primitive
     * @param {?} entity
     * @return {?}
     */
    function (id, primitive, entity) {
        if (!this._views.has(id) && primitive.show) {
            /** @type {?} */
            var context = new AcHtmlContext(id, { $implicit: entity });
            /** @type {?} */
            var viewRef = this._viewContainerRef.createEmbeddedView(this._templateRef, context);
            this._views.set(id, { viewRef: viewRef, context: context });
            this._changeDetector.detectChanges();
        }
        else if (this._views.has(id) && !primitive.show) {
            this.remove(id, primitive);
        }
        else if (this._views.has(id) && primitive.show) {
            this._changeDetector.detectChanges();
        }
    };
    /**
     * @param {?} id
     * @param {?} primitive
     * @return {?}
     */
    AcHtmlDirective.prototype.addOrUpdate = /**
     * @param {?} id
     * @param {?} primitive
     * @return {?}
     */
    function (id, primitive) {
        /** @type {?} */
        var context = this._layerService.context;
        /** @type {?} */
        var entity = context[this._layerService.getEntityName()];
        if (this._views.has(id)) {
            this._views.get(id).context.context.$implicit = entity;
        }
        this._acHtmlManager.addOrUpdate(id, { entity: entity, primitive: primitive });
        this._handleView(id, primitive, entity);
    };
    /**
     * @param {?} id
     * @param {?} primitive
     * @return {?}
     */
    AcHtmlDirective.prototype.remove = /**
     * @param {?} id
     * @param {?} primitive
     * @return {?}
     */
    function (id, primitive) {
        if (!this._views.has(id)) {
            return;
        }
        var viewRef = this._views.get(id).viewRef;
        this._viewContainerRef.remove(this._viewContainerRef.indexOf(viewRef));
        this._views.delete(id);
        this._acHtmlManager.remove(id);
        primitive.element = null;
    };
    AcHtmlDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[acHtml]',
                },] }
    ];
    /** @nocollapse */
    AcHtmlDirective.ctorParameters = function () { return [
        { type: TemplateRef },
        { type: ViewContainerRef },
        { type: ChangeDetectorRef },
        { type: LayerService },
        { type: AcHtmlManager }
    ]; };
    return AcHtmlDirective;
}());
export { AcHtmlDirective };
if (false) {
    /**
     * @type {?}
     * @private
     */
    AcHtmlDirective.prototype._views;
    /**
     * @type {?}
     * @private
     */
    AcHtmlDirective.prototype._templateRef;
    /**
     * @type {?}
     * @private
     */
    AcHtmlDirective.prototype._viewContainerRef;
    /**
     * @type {?}
     * @private
     */
    AcHtmlDirective.prototype._changeDetector;
    /**
     * @type {?}
     * @private
     */
    AcHtmlDirective.prototype._layerService;
    /**
     * @type {?}
     * @private
     */
    AcHtmlDirective.prototype._acHtmlManager;
}
