// angular-cesium/esm5/lib/angular-cesium-widgets/components/toolbar/ac-toolbar-button/ac-toolbar-button.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, EventEmitter, Input, Output } from '@angular/core';
/**
 * Toolbar button widget, act as a single button inside ac-toolbar component
 * Can accepts content components or passing [iconUrl]
 * configure with: `[iconUrl]`,`[buttonClass]`,`[iconClass]`,`(onClick)`
 *
 * Usage:
 * ```
 * <ac-toolbar [allowDrag]="true">
 * <ac-toolbar-button [iconUrl]="'assets/home-icon.svg'" (onClick)="goHome()">
 * </ac-toolbar-button>
 * <ac-toolbar-button [iconUrl]="'assets/explore-icon.svg'" (onClick)="rangeAndBearing()">
 * </ac-toolbar-button>
 * </ac-toolbar>
 * ```
 *
 */
var AcToolbarButtonComponent = /** @class */ (function () {
    function AcToolbarButtonComponent() {
        this.onClick = new EventEmitter();
    }
    /**
     * @return {?}
     */
    AcToolbarButtonComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    AcToolbarButtonComponent.decorators = [
        { type: Component, args: [{
                    selector: 'ac-toolbar-button',
                    template: "\n        <div (click)=\"onClick.emit()\" class=\"button-container {{buttonClass}}\">\n            <img *ngIf=\"iconUrl\" [src]=\"iconUrl\" class=\"icon {{iconClass}}\"/>\n            <ng-content></ng-content>\n        </div>\n    ",
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    styles: ["\n        .button-container {\n            border-radius: 1px;\n            background-color: rgba(255, 255, 255, 0.8);\n            height: 30px;\n            width: 30px;\n            padding: 5px;\n            transition: all 0.2s;\n            cursor: pointer;\n            display: flex;\n            justify-content: center;\n            align-items: center;\n            flex-direction: column;\n        }\n\n        .button-container:hover {\n            background-color: rgba(255, 255, 255, 0.95);\n        }\n\n        .icon {\n            height: 30px;\n            width: 30px;\n        }\n    "]
                }] }
    ];
    /** @nocollapse */
    AcToolbarButtonComponent.ctorParameters = function () { return []; };
    AcToolbarButtonComponent.propDecorators = {
        iconUrl: [{ type: Input }],
        buttonClass: [{ type: Input }],
        iconClass: [{ type: Input }],
        onClick: [{ type: Output }]
    };
    return AcToolbarButtonComponent;
}());
export { AcToolbarButtonComponent };
if (false) {
    /** @type {?} */
    AcToolbarButtonComponent.prototype.iconUrl;
    /** @type {?} */
    AcToolbarButtonComponent.prototype.buttonClass;
    /** @type {?} */
    AcToolbarButtonComponent.prototype.iconClass;
    /** @type {?} */
    AcToolbarButtonComponent.prototype.onClick;
}
