// angular-cesium/esm5/lib/angular-cesium/services/basic-primitive-desc/basic-static-primitive-desc.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Input } from '@angular/core';
import { BasicDesc } from '../basic-desc/basic-desc.service';
var BasicStaticPrimitiveDesc = /** @class */ (function (_super) {
    tslib_1.__extends(BasicStaticPrimitiveDesc, _super);
    function BasicStaticPrimitiveDesc(_staticPrimitiveDrawer, layerService, computationCache, cesiumProperties) {
        var _this = _super.call(this, _staticPrimitiveDrawer, layerService, computationCache, cesiumProperties) || this;
        _this._staticPrimitiveDrawer = _staticPrimitiveDrawer;
        return _this;
    }
    /**
     * @return {?}
     */
    BasicStaticPrimitiveDesc.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this._layerService.registerDescription(this);
        this._geometryPropsEvaluator = this._cesiumProperties.createEvaluator(this.geometryProps);
        this._instancePropsEvaluator = this._cesiumProperties.createEvaluator(this.instanceProps);
        this._primitivePropsEvaluator = this._cesiumProperties.createEvaluator(this.primitiveProps);
    };
    /**
     * @param {?} context
     * @param {?} id
     * @param {?} entity
     * @return {?}
     */
    BasicStaticPrimitiveDesc.prototype.draw = /**
     * @param {?} context
     * @param {?} id
     * @param {?} entity
     * @return {?}
     */
    function (context, id, entity) {
        /** @type {?} */
        var geometryProps = this._geometryPropsEvaluator(this._computationCache, context);
        /** @type {?} */
        var instanceProps = this._instancePropsEvaluator(this._computationCache, context);
        /** @type {?} */
        var primitiveProps = this._primitivePropsEvaluator(this._computationCache, context);
        if (!this._cesiumObjectsMap.has(id)) {
            /** @type {?} */
            var primitive = this._staticPrimitiveDrawer.add(geometryProps, instanceProps, primitiveProps);
            primitive.acEntity = entity; // set the entity on the primitive for later usage
            this._cesiumObjectsMap.set(id, primitive);
        }
        else {
            /** @type {?} */
            var primitive = this._cesiumObjectsMap.get(id);
            this._staticPrimitiveDrawer.update(primitive, geometryProps, instanceProps, primitiveProps);
        }
    };
    BasicStaticPrimitiveDesc.propDecorators = {
        geometryProps: [{ type: Input }],
        instanceProps: [{ type: Input }],
        primitiveProps: [{ type: Input }]
    };
    return BasicStaticPrimitiveDesc;
}(BasicDesc));
export { BasicStaticPrimitiveDesc };
if (false) {
    /** @type {?} */
    BasicStaticPrimitiveDesc.prototype.geometryProps;
    /** @type {?} */
    BasicStaticPrimitiveDesc.prototype.instanceProps;
    /** @type {?} */
    BasicStaticPrimitiveDesc.prototype.primitiveProps;
    /**
     * @type {?}
     * @private
     */
    BasicStaticPrimitiveDesc.prototype._geometryPropsEvaluator;
    /**
     * @type {?}
     * @private
     */
    BasicStaticPrimitiveDesc.prototype._instancePropsEvaluator;
    /**
     * @type {?}
     * @private
     */
    BasicStaticPrimitiveDesc.prototype._primitivePropsEvaluator;
    /**
     * @type {?}
     * @protected
     */
    BasicStaticPrimitiveDesc.prototype._staticPrimitiveDrawer;
}
