// angular-cesium/esm2015/lib/angular-cesium/services/selection-manager/selection-manager.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { filter, map } from 'rxjs/operators';
import { Injectable } from '@angular/core';
import { BehaviorSubject, Subject } from 'rxjs';
import { CesiumEvent } from '../map-events-mananger/consts/cesium-event.enum';
import { PickOptions } from '../map-events-mananger/consts/pickOptions.enum';
import { MapsManagerService } from '../maps-manager/maps-manager.service';
/**
 * @record
 */
export function SelectionOptions() { }
if (false) {
    /** @type {?|undefined} */
    SelectionOptions.prototype.event;
    /** @type {?|undefined} */
    SelectionOptions.prototype.modifier;
    /** @type {?|undefined} */
    SelectionOptions.prototype.entityType;
}
/**
 * Manages entity selection service for any given mouse event and modifier
 * the service will manage the list of selected items.
 * check out the example
 * you must provide the service yourself
 *
 *  __Usage :__
 * ```
 * // provide the service in some component
 * \@Component({
 * //...
 *  providers: [SelectionManagerService]
 * }
 *
 * // Usage example:
 * // init selection
 * const selectedIndicator = ture; // optional default true, if true a boolean "selected" property will be added to the selected entity
 * selectionManagerService.initSelection({ event: CesiumEvent.LEFT_CLICK,
 * 																			modifier: CesiumEventModifier.CTRL
 * 																		},selectedIndicator);
 * // Get selected
 * const selected = selectionManagerService.selected();
 *
 * // Or as observer
 * const selected$ = selectionManagerService.selected$();
 *
 * ```
 *
 */
export class SelectionManagerService {
    /**
     * @param {?} mapsManager
     */
    constructor(mapsManager) {
        this.mapsManager = mapsManager;
        this.selectedEntitiesItems$ = new BehaviorSubject([]);
        this.selectedEntitySubject$ = new Subject();
    }
    /**
     * @return {?}
     */
    selectedEntities$() {
        return this.selectedEntitiesItems$.asObservable();
    }
    /**
     * @return {?}
     */
    selectedEntities() {
        return this.selectedEntitiesItems$.getValue();
    }
    /**
     * @return {?}
     */
    selectedEntity$() {
        return this.selectedEntitySubject$;
    }
    /**
     * @param {?} entity
     * @param {?} addSelectedIndicator
     * @return {?}
     */
    toggleSelection(entity, addSelectedIndicator) {
        /** @type {?} */
        const current = this.selectedEntities();
        if (current.indexOf(entity) === -1) {
            this.addToSelected(entity, addSelectedIndicator);
        }
        else {
            this.removeSelected(entity, addSelectedIndicator);
        }
    }
    /**
     * @private
     * @param {?} entity
     * @param {?} addSelectedIndicator
     * @return {?}
     */
    addToSelected(entity, addSelectedIndicator) {
        if (addSelectedIndicator) {
            entity['selected'] = true;
        }
        /** @type {?} */
        const current = this.selectedEntities();
        this.selectedEntitySubject$.next(entity);
        this.selectedEntitiesItems$.next([...current, entity]);
    }
    /**
     * @private
     * @param {?} entity
     * @param {?} addSelectedIndicator
     * @return {?}
     */
    removeSelected(entity, addSelectedIndicator) {
        if (addSelectedIndicator) {
            entity['selected'] = false;
        }
        /** @type {?} */
        const current = this.selectedEntities();
        /** @type {?} */
        const entityIndex = current.indexOf(entity);
        if (entityIndex !== -1) {
            current.splice(entityIndex, 1);
            this.selectedEntitiesItems$.next(current);
            this.selectedEntitySubject$.next(entity);
        }
    }
    /**
     * @param {?=} selectionOptions
     * @param {?=} addSelectedIndicator
     * @param {?=} eventPriority
     * @param {?=} mapId
     * @return {?}
     */
    initSelection(selectionOptions, addSelectedIndicator = true, eventPriority, mapId) {
        /** @type {?} */
        const mapComponent = this.mapsManager.getMap(mapId);
        if (!mapComponent) {
            return;
        }
        this.mapEventsManagerService = mapComponent.getMapEventsManager();
        if (!selectionOptions) {
            Object.assign(selectionOptions, { event: CesiumEvent.LEFT_CLICK });
        }
        /** @type {?} */
        const eventSubscription = this.mapEventsManagerService.register({
            event: selectionOptions.event,
            pick: PickOptions.PICK_ONE,
            modifier: selectionOptions.modifier,
            entityType: selectionOptions.entityType,
            priority: eventPriority,
        });
        eventSubscription.pipe(map(result => result.entities), filter(entities => entities && entities.length > 0))
            .subscribe(entities => {
            /** @type {?} */
            const entity = entities[0];
            this.toggleSelection(entity, addSelectedIndicator);
        });
    }
}
SelectionManagerService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
SelectionManagerService.ctorParameters = () => [
    { type: MapsManagerService }
];
if (false) {
    /** @type {?} */
    SelectionManagerService.prototype.selectedEntitiesItems$;
    /** @type {?} */
    SelectionManagerService.prototype.selectedEntitySubject$;
    /**
     * @type {?}
     * @private
     */
    SelectionManagerService.prototype.mapEventsManagerService;
    /**
     * @type {?}
     * @private
     */
    SelectionManagerService.prototype.mapsManager;
}
