// angular-cesium/esm2015/lib/angular-cesium/models/ac-keyboard-action.enum.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @enum {number} */
const KeyboardAction = {
    CAMERA_FORWARD: 0,
    CAMERA_BACKWARD: 1,
    CAMERA_RIGHT: 2,
    CAMERA_LEFT: 3,
    CAMERA_UP: 4,
    CAMERA_DOWN: 5,
    CAMERA_LOOK_RIGHT: 6,
    CAMERA_LOOK_LEFT: 7,
    CAMERA_LOOK_UP: 8,
    CAMERA_LOOK_DOWN: 9,
    CAMERA_TWIST_RIGHT: 10,
    CAMERA_TWIST_LEFT: 11,
    CAMERA_ROTATE_RIGHT: 12,
    CAMERA_ROTATE_LEFT: 13,
    CAMERA_ROTATE_UP: 14,
    CAMERA_ROTATE_DOWN: 15,
    CAMERA_ZOOM_IN: 16,
    CAMERA_ZOOM_OUT: 17,
};
export { KeyboardAction };
KeyboardAction[KeyboardAction.CAMERA_FORWARD] = 'CAMERA_FORWARD';
KeyboardAction[KeyboardAction.CAMERA_BACKWARD] = 'CAMERA_BACKWARD';
KeyboardAction[KeyboardAction.CAMERA_RIGHT] = 'CAMERA_RIGHT';
KeyboardAction[KeyboardAction.CAMERA_LEFT] = 'CAMERA_LEFT';
KeyboardAction[KeyboardAction.CAMERA_UP] = 'CAMERA_UP';
KeyboardAction[KeyboardAction.CAMERA_DOWN] = 'CAMERA_DOWN';
KeyboardAction[KeyboardAction.CAMERA_LOOK_RIGHT] = 'CAMERA_LOOK_RIGHT';
KeyboardAction[KeyboardAction.CAMERA_LOOK_LEFT] = 'CAMERA_LOOK_LEFT';
KeyboardAction[KeyboardAction.CAMERA_LOOK_UP] = 'CAMERA_LOOK_UP';
KeyboardAction[KeyboardAction.CAMERA_LOOK_DOWN] = 'CAMERA_LOOK_DOWN';
KeyboardAction[KeyboardAction.CAMERA_TWIST_RIGHT] = 'CAMERA_TWIST_RIGHT';
KeyboardAction[KeyboardAction.CAMERA_TWIST_LEFT] = 'CAMERA_TWIST_LEFT';
KeyboardAction[KeyboardAction.CAMERA_ROTATE_RIGHT] = 'CAMERA_ROTATE_RIGHT';
KeyboardAction[KeyboardAction.CAMERA_ROTATE_LEFT] = 'CAMERA_ROTATE_LEFT';
KeyboardAction[KeyboardAction.CAMERA_ROTATE_UP] = 'CAMERA_ROTATE_UP';
KeyboardAction[KeyboardAction.CAMERA_ROTATE_DOWN] = 'CAMERA_ROTATE_DOWN';
KeyboardAction[KeyboardAction.CAMERA_ZOOM_IN] = 'CAMERA_ZOOM_IN';
KeyboardAction[KeyboardAction.CAMERA_ZOOM_OUT] = 'CAMERA_ZOOM_OUT';
