// angular-cesium/esm2015/lib/angular-cesium/services/map-layers/map-layers.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { CesiumService } from '../cesium/cesium.service';
import { Injectable } from '@angular/core';
export class MapLayersService {
    /**
     * @param {?} cesiumService
     */
    constructor(cesiumService) {
        this.cesiumService = cesiumService;
        this.layersDataSources = [];
    }
    /**
     * @param {?} dataSources
     * @param {?} zIndex
     * @return {?}
     */
    registerLayerDataSources(dataSources, zIndex) {
        dataSources.forEach(ds => {
            ds.zIndex = zIndex;
            this.layersDataSources.push(ds);
        });
    }
    /**
     * @return {?}
     */
    drawAllLayers() {
        this.layersDataSources.sort((a, b) => a.zIndex - b.zIndex);
        this.layersDataSources.forEach((dataSource) => {
            this.cesiumService.getViewer().dataSources.add(dataSource);
        });
    }
    /**
     * @param {?} dataSources
     * @param {?} newZIndex
     * @return {?}
     */
    updateAndRefresh(dataSources, newZIndex) {
        if (dataSources && dataSources.length) {
            dataSources.forEach((ds) => {
                /** @type {?} */
                const index = this.layersDataSources.indexOf(ds);
                if (index !== -1) {
                    this.layersDataSources[index].zIndex = newZIndex;
                }
            });
            this.cesiumService.getViewer().dataSources.removeAll();
            this.drawAllLayers();
        }
    }
    /**
     * @param {?} dataSources
     * @return {?}
     */
    removeDataSources(dataSources) {
        dataSources.forEach(ds => {
            /** @type {?} */
            const index = this.layersDataSources.indexOf(ds);
            if (index !== -1) {
                this.layersDataSources.splice(index, 1);
                this.cesiumService.getViewer().dataSources.remove(ds, true);
            }
        });
    }
}
MapLayersService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
MapLayersService.ctorParameters = () => [
    { type: CesiumService }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    MapLayersService.prototype.layersDataSources;
    /**
     * @type {?}
     * @private
     */
    MapLayersService.prototype.cesiumService;
}
