// angular-cesium/esm5/lib/angular-cesium/services/drawers/czml-drawer/czml-drawer.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { CesiumService } from '../../cesium/cesium.service';
import { BasicDrawerService } from '../basic-drawer/basic-drawer.service';
/**
 *  This drawer is responsible for drawing czml dataSources.
 */
var CzmlDrawerService = /** @class */ (function (_super) {
    tslib_1.__extends(CzmlDrawerService, _super);
    function CzmlDrawerService(cesiumService) {
        var _this = _super.call(this) || this;
        _this.cesiumService = cesiumService;
        return _this;
    }
    /**
     * @param {?=} options
     * @return {?}
     */
    CzmlDrawerService.prototype.init = /**
     * @param {?=} options
     * @return {?}
     */
    function (options) {
        /** @type {?} */
        var dataSources = [];
        this.czmlStream = new Cesium.CzmlDataSource('czml');
        dataSources.push(this.czmlStream);
        this.cesiumService.getViewer().dataSources.add(this.czmlStream);
        return dataSources;
    };
    // returns the packet, provided by the stream
    // returns the packet, provided by the stream
    /**
     * @param {?} cesiumProps
     * @return {?}
     */
    CzmlDrawerService.prototype.add = 
    // returns the packet, provided by the stream
    /**
     * @param {?} cesiumProps
     * @return {?}
     */
    function (cesiumProps) {
        this.czmlStream.process(cesiumProps.czmlPacket);
        return cesiumProps;
    };
    /**
     * @param {?} entity
     * @param {?} cesiumProps
     * @return {?}
     */
    CzmlDrawerService.prototype.update = /**
     * @param {?} entity
     * @param {?} cesiumProps
     * @return {?}
     */
    function (entity, cesiumProps) {
        this.czmlStream.process(cesiumProps.czmlPacket);
    };
    /**
     * @param {?} entity
     * @return {?}
     */
    CzmlDrawerService.prototype.remove = /**
     * @param {?} entity
     * @return {?}
     */
    function (entity) {
        this.czmlStream.entities.removeById(entity.acEntity.id);
    };
    /**
     * @return {?}
     */
    CzmlDrawerService.prototype.removeAll = /**
     * @return {?}
     */
    function () {
        this.czmlStream.entities.removeAll();
    };
    /**
     * @param {?} showValue
     * @return {?}
     */
    CzmlDrawerService.prototype.setShow = /**
     * @param {?} showValue
     * @return {?}
     */
    function (showValue) {
        this.czmlStream.entities.show = showValue;
    };
    CzmlDrawerService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    CzmlDrawerService.ctorParameters = function () { return [
        { type: CesiumService }
    ]; };
    return CzmlDrawerService;
}(BasicDrawerService));
export { CzmlDrawerService };
if (false) {
    /** @type {?} */
    CzmlDrawerService.prototype.czmlStream;
    /**
     * @type {?}
     * @private
     */
    CzmlDrawerService.prototype.cesiumService;
}
