// angular-cesium/esm2015/lib/angular-cesium/services/plonter/plonter.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { EventEmitter, Injectable } from '@angular/core';
import { Subject } from 'rxjs';
/**
 * Service for solving plonter.
 * Used by map-event-manager and plonter context component
 */
export class PlonterService {
    constructor() {
        this._entitesToPlonter = [];
        this._plonterChangeNotifier = new EventEmitter();
        this._plonterObserver = new Subject();
    }
    /**
     * @return {?}
     */
    get plonterChangeNotifier() {
        return this._plonterChangeNotifier;
    }
    /**
     * @return {?}
     */
    get plonterShown() {
        return this._plonterShown;
    }
    /**
     * @return {?}
     */
    get entitesToPlonter() {
        return this._entitesToPlonter;
    }
    /**
     * @return {?}
     */
    get plonterClickPosition() {
        return this._eventResult.movement;
    }
    /**
     * @param {?} eventResult
     * @return {?}
     */
    plonterIt(eventResult) {
        this._eventResult = eventResult;
        this._entitesToPlonter = eventResult.entities;
        this._plonterShown = true;
        this._plonterChangeNotifier.emit();
        return this._plonterObserver;
    }
    /**
     * @param {?} entity
     * @return {?}
     */
    resolvePlonter(entity) {
        this._plonterShown = false;
        this._eventResult.entities = [entity];
        this._plonterChangeNotifier.emit();
        this._plonterObserver.next(this._eventResult);
    }
}
PlonterService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
PlonterService.ctorParameters = () => [];
if (false) {
    /**
     * @type {?}
     * @private
     */
    PlonterService.prototype._plonterShown;
    /**
     * @type {?}
     * @private
     */
    PlonterService.prototype._entitesToPlonter;
    /**
     * @type {?}
     * @private
     */
    PlonterService.prototype._plonterObserver;
    /**
     * @type {?}
     * @private
     */
    PlonterService.prototype._eventResult;
    /**
     * @type {?}
     * @private
     */
    PlonterService.prototype._plonterChangeNotifier;
}
