// angular-cesium/esm2015/lib/angular-cesium/services/cesium-properties/cesium-properties.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { JsonMapper } from '../json-mapper/json-mapper.service';
import { Parse } from 'angular2parse';
import { SmartAssigner } from '../smart-assigner/smart-assigner.service';
export class CesiumProperties {
    /**
     * @param {?} _parser
     * @param {?} _jsonMapper
     */
    constructor(_parser, _jsonMapper) {
        this._parser = _parser;
        this._jsonMapper = _jsonMapper;
        this._assignersCache = new Map();
        this._evaluatorsCache = new Map();
    }
    /**
     * @param {?} expression
     * @param {?=} withCache
     * @return {?}
     */
    _compile(expression, withCache = true) {
        /** @type {?} */
        const cesiumDesc = {};
        /** @type {?} */
        const propsMap = new Map();
        /** @type {?} */
        const resultMap = this._jsonMapper.map(expression);
        resultMap.forEach((resultExpression, prop) => propsMap.set(prop, {
            expression: resultExpression,
            get: this._parser.eval(resultExpression)
        }));
        propsMap.forEach((value, prop) => {
            if (withCache) {
                cesiumDesc[prop || 'undefined'] = `cache.get(\`${value.expression}\`, () => propsMap.get('${prop}').get(context))`;
            }
            else {
                cesiumDesc[prop || 'undefined'] = `propsMap.get('${prop}').get(context)`;
            }
        });
        /** @type {?} */
        const fnBody = `return ${JSON.stringify(cesiumDesc).replace(/"/g, '')};`;
        /** @type {?} */
        const getFn = new Function('propsMap', 'cache', 'context', fnBody);
        return function evaluateCesiumProps(cache, context) {
            return getFn(propsMap, cache, context);
        };
    }
    /**
     * @param {?} expression
     * @return {?}
     */
    _build(expression) {
        /** @type {?} */
        const props = Array.from(this._jsonMapper.map(expression).keys());
        /** @type {?} */
        const smartAssigner = SmartAssigner.create(props);
        return function assignCesiumProps(oldVal, newVal) {
            return smartAssigner(oldVal, newVal);
        };
    }
    /**
     * @param {?} expression
     * @param {?=} withCache
     * @param {?=} newEvaluator
     * @return {?}
     */
    createEvaluator(expression, withCache = true, newEvaluator = false) {
        if (!newEvaluator && this._evaluatorsCache.has(expression)) {
            return this._evaluatorsCache.get(expression);
        }
        /** @type {?} */
        const evaluatorFn = this._compile(expression, withCache);
        this._evaluatorsCache.set(expression, evaluatorFn);
        return evaluatorFn;
    }
    /**
     * @param {?} expression
     * @return {?}
     */
    createAssigner(expression) {
        if (this._assignersCache.has(expression)) {
            return this._assignersCache.get(expression);
        }
        /** @type {?} */
        const assignFn = this._build(expression);
        this._assignersCache.set(expression, assignFn);
        return assignFn;
    }
}
CesiumProperties.decorators = [
    { type: Injectable }
];
/** @nocollapse */
CesiumProperties.ctorParameters = () => [
    { type: Parse },
    { type: JsonMapper }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    CesiumProperties.prototype._assignersCache;
    /**
     * @type {?}
     * @private
     */
    CesiumProperties.prototype._evaluatorsCache;
    /**
     * @type {?}
     * @private
     */
    CesiumProperties.prototype._parser;
    /**
     * @type {?}
     * @private
     */
    CesiumProperties.prototype._jsonMapper;
}
