// angular-cesium/esm5/lib/angular-cesium/services/cesium/cesium.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, NgZone, Optional } from '@angular/core';
import { ViewerFactory } from '../viewer-factory/viewer-factory.service';
import { ViewerConfiguration } from '../viewer-configuration/viewer-configuration.service';
/**
 *  Service that initialize cesium viewer and expose cesium viewer and scene.
 */
var CesiumService = /** @class */ (function () {
    function CesiumService(ngZone, viewerFactory, viewerConfiguration) {
        this.ngZone = ngZone;
        this.viewerFactory = viewerFactory;
        this.viewerConfiguration = viewerConfiguration;
    }
    /**
     * @param {?} mapContainer
     * @param {?} map
     * @return {?}
     */
    CesiumService.prototype.init = /**
     * @param {?} mapContainer
     * @param {?} map
     * @return {?}
     */
    function (mapContainer, map) {
        var _this = this;
        this.map = map;
        this.ngZone.runOutsideAngular(function () {
            /** @type {?} */
            var options = _this.viewerConfiguration ? _this.viewerConfiguration.getNextViewerOptions() : undefined;
            _this.cesiumViewer = _this.viewerFactory.createViewer(mapContainer, options);
            /** @type {?} */
            var viewerModifier = _this.viewerConfiguration && _this.viewerConfiguration.getNextViewerModifier();
            if (typeof viewerModifier === 'function') {
                viewerModifier(_this.cesiumViewer);
            }
        });
    };
    /**
     * For more information see https://cesiumjs.org/Cesium/Build/Documentation/Viewer.html?classFilter=viewe
     * @returns cesiumViewer
     */
    /**
     * For more information see https://cesiumjs.org/Cesium/Build/Documentation/Viewer.html?classFilter=viewe
     * @return {?} cesiumViewer
     */
    CesiumService.prototype.getViewer = /**
     * For more information see https://cesiumjs.org/Cesium/Build/Documentation/Viewer.html?classFilter=viewe
     * @return {?} cesiumViewer
     */
    function () {
        return this.cesiumViewer;
    };
    /**
     * For more information see https://cesiumjs.org/Cesium/Build/Documentation/Scene.html?classFilter=scene
     * @returns cesium scene
     */
    /**
     * For more information see https://cesiumjs.org/Cesium/Build/Documentation/Scene.html?classFilter=scene
     * @return {?} cesium scene
     */
    CesiumService.prototype.getScene = /**
     * For more information see https://cesiumjs.org/Cesium/Build/Documentation/Scene.html?classFilter=scene
     * @return {?} cesium scene
     */
    function () {
        return this.cesiumViewer.scene;
    };
    /**
     * For more information see https://developer.mozilla.org/en-US/docs/Web/API/Canvas_API
     * @returns cesium canvas
     */
    /**
     * For more information see https://developer.mozilla.org/en-US/docs/Web/API/Canvas_API
     * @return {?} cesium canvas
     */
    CesiumService.prototype.getCanvas = /**
     * For more information see https://developer.mozilla.org/en-US/docs/Web/API/Canvas_API
     * @return {?} cesium canvas
     */
    function () {
        return (/** @type {?} */ (this.cesiumViewer.canvas));
    };
    /**
     * @return {?}
     */
    CesiumService.prototype.getMap = /**
     * @return {?}
     */
    function () {
        return this.map;
    };
    CesiumService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    CesiumService.ctorParameters = function () { return [
        { type: NgZone },
        { type: ViewerFactory },
        { type: ViewerConfiguration, decorators: [{ type: Optional }] }
    ]; };
    return CesiumService;
}());
export { CesiumService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    CesiumService.prototype.cesiumViewer;
    /**
     * @type {?}
     * @private
     */
    CesiumService.prototype.map;
    /**
     * @type {?}
     * @private
     */
    CesiumService.prototype.ngZone;
    /**
     * @type {?}
     * @private
     */
    CesiumService.prototype.viewerFactory;
    /**
     * @type {?}
     * @private
     */
    CesiumService.prototype.viewerConfiguration;
}
