// angular-cesium/esm5/lib/angular-cesium/services/map-events-mananger/cesium-event-builder.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { publish } from 'rxjs/operators';
import { CesiumService } from '../cesium/cesium.service';
import { CesiumEvent } from './consts/cesium-event.enum';
import { Injectable } from '@angular/core';
import { CesiumPureEventObserver } from './event-observers/cesium-pure-event-observer';
import { CesiumLongPressObserver } from './event-observers/cesium-long-press-observer';
var CesiumEventBuilder = /** @class */ (function () {
    function CesiumEventBuilder(cesiumService) {
        this.cesiumService = cesiumService;
        this.cesiumEventsObservables = new Map();
    }
    /**
     * @param {?} event
     * @param {?=} modifier
     * @return {?}
     */
    CesiumEventBuilder.getEventFullName = /**
     * @param {?} event
     * @param {?=} modifier
     * @return {?}
     */
    function (event, modifier) {
        if (modifier) {
            return event + "_" + modifier;
        }
        else {
            return event.toString();
        }
    };
    /**
     * @return {?}
     */
    CesiumEventBuilder.prototype.init = /**
     * @return {?}
     */
    function () {
        this.eventsHandler = this.cesiumService.getViewer().screenSpaceEventHandler;
    };
    /**
     * @param {?} event
     * @param {?=} modifier
     * @return {?}
     */
    CesiumEventBuilder.prototype.get = /**
     * @param {?} event
     * @param {?=} modifier
     * @return {?}
     */
    function (event, modifier) {
        /** @type {?} */
        var eventName = CesiumEventBuilder.getEventFullName(event, modifier);
        if (this.cesiumEventsObservables.has(eventName)) {
            return this.cesiumEventsObservables.get(eventName);
        }
        else {
            /** @type {?} */
            var eventObserver = this.createCesiumEventObservable(event, modifier);
            this.cesiumEventsObservables.set(eventName, eventObserver);
            return eventObserver;
        }
    };
    /**
     * @private
     * @param {?} event
     * @param {?=} modifier
     * @return {?}
     */
    CesiumEventBuilder.prototype.createCesiumEventObservable = /**
     * @private
     * @param {?} event
     * @param {?=} modifier
     * @return {?}
     */
    function (event, modifier) {
        /** @type {?} */
        var cesiumEventObservable;
        if (CesiumEventBuilder.longPressEvents.has(event)) {
            cesiumEventObservable = this.createSpecialCesiumEventObservable(event, modifier);
        }
        else {
            cesiumEventObservable = publish()(new CesiumPureEventObserver(event, modifier).init(this.eventsHandler));
        }
        cesiumEventObservable.connect();
        return cesiumEventObservable;
    };
    /**
     * @private
     * @param {?} event
     * @param {?} modifier
     * @return {?}
     */
    CesiumEventBuilder.prototype.createSpecialCesiumEventObservable = /**
     * @private
     * @param {?} event
     * @param {?} modifier
     * @return {?}
     */
    function (event, modifier) {
        // could support more events if needed
        return new CesiumLongPressObserver(event, modifier, this).init();
    };
    CesiumEventBuilder.longPressEvents = new Set([
        CesiumEvent.LONG_LEFT_PRESS,
        CesiumEvent.LONG_RIGHT_PRESS,
        CesiumEvent.LONG_MIDDLE_PRESS
    ]);
    CesiumEventBuilder.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    CesiumEventBuilder.ctorParameters = function () { return [
        { type: CesiumService }
    ]; };
    return CesiumEventBuilder;
}());
export { CesiumEventBuilder };
if (false) {
    /** @type {?} */
    CesiumEventBuilder.longPressEvents;
    /**
     * @type {?}
     * @private
     */
    CesiumEventBuilder.prototype.eventsHandler;
    /**
     * @type {?}
     * @private
     */
    CesiumEventBuilder.prototype.cesiumEventsObservables;
    /**
     * @type {?}
     * @private
     */
    CesiumEventBuilder.prototype.cesiumService;
}
