// angular-cesium/esm5/lib/angular-cesium/services/drawers/primitives-drawer/primitives-drawer.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { BasicDrawerService } from '../basic-drawer/basic-drawer.service';
/**
 *  General primitives drawer responsible of drawing Cesium primitives.
 *  Drawers the handle Cesium primitives extend it.
 * @abstract
 */
var /**
 *  General primitives drawer responsible of drawing Cesium primitives.
 *  Drawers the handle Cesium primitives extend it.
 * @abstract
 */
PrimitivesDrawerService = /** @class */ (function (_super) {
    tslib_1.__extends(PrimitivesDrawerService, _super);
    function PrimitivesDrawerService(drawerType, cesiumService) {
        var _this = _super.call(this) || this;
        _this.drawerType = drawerType;
        _this.cesiumService = cesiumService;
        _this._show = true;
        return _this;
    }
    /**
     * @return {?}
     */
    PrimitivesDrawerService.prototype.init = /**
     * @return {?}
     */
    function () {
        this._cesiumCollection = new this.drawerType();
        this._primitiveCollectionWrap = new Cesium.PrimitiveCollection();
        this._primitiveCollectionWrap.add(this._cesiumCollection);
        this.cesiumService.getScene().primitives.add(this._primitiveCollectionWrap);
    };
    /**
     * @param {?} cesiumProps
     * @param {...?} args
     * @return {?}
     */
    PrimitivesDrawerService.prototype.add = /**
     * @param {?} cesiumProps
     * @param {...?} args
     * @return {?}
     */
    function (cesiumProps) {
        var args = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            args[_i - 1] = arguments[_i];
        }
        return this._cesiumCollection.add(cesiumProps);
    };
    /**
     * @param {?} entity
     * @param {?} cesiumProps
     * @param {...?} args
     * @return {?}
     */
    PrimitivesDrawerService.prototype.update = /**
     * @param {?} entity
     * @param {?} cesiumProps
     * @param {...?} args
     * @return {?}
     */
    function (entity, cesiumProps) {
        var args = [];
        for (var _i = 2; _i < arguments.length; _i++) {
            args[_i - 2] = arguments[_i];
        }
        if (this._propsAssigner) {
            this._propsAssigner(entity, cesiumProps);
        }
        else {
            Object.assign(entity, cesiumProps);
        }
    };
    /**
     * @param {?} entity
     * @return {?}
     */
    PrimitivesDrawerService.prototype.remove = /**
     * @param {?} entity
     * @return {?}
     */
    function (entity) {
        this._cesiumCollection.remove(entity);
    };
    /**
     * @return {?}
     */
    PrimitivesDrawerService.prototype.removeAll = /**
     * @return {?}
     */
    function () {
        this._cesiumCollection.removeAll();
    };
    /**
     * @param {?} showValue
     * @return {?}
     */
    PrimitivesDrawerService.prototype.setShow = /**
     * @param {?} showValue
     * @return {?}
     */
    function (showValue) {
        this._show = showValue;
        this._primitiveCollectionWrap.show = showValue;
    };
    /**
     * @return {?}
     */
    PrimitivesDrawerService.prototype.getShow = /**
     * @return {?}
     */
    function () {
        return this._show;
    };
    return PrimitivesDrawerService;
}(BasicDrawerService));
/**
 *  General primitives drawer responsible of drawing Cesium primitives.
 *  Drawers the handle Cesium primitives extend it.
 * @abstract
 */
export { PrimitivesDrawerService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    PrimitivesDrawerService.prototype._show;
    /**
     * @type {?}
     * @private
     */
    PrimitivesDrawerService.prototype._primitiveCollectionWrap;
    /**
     * @type {?}
     * @protected
     */
    PrimitivesDrawerService.prototype._cesiumCollection;
    /**
     * @type {?}
     * @protected
     */
    PrimitivesDrawerService.prototype._propsAssigner;
    /**
     * @type {?}
     * @private
     */
    PrimitivesDrawerService.prototype.drawerType;
    /**
     * @type {?}
     * @private
     */
    PrimitivesDrawerService.prototype.cesiumService;
}
