// angular-cesium/esm5/lib/angular-cesium/services/drawers/basic-drawer/basic-drawer.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 *  Abstract drawer. All drawers extends this class.
 */
/**
 * @abstract
 */
var /**
 * @abstract
 */
BasicDrawerService = /** @class */ (function () {
    function BasicDrawerService() {
    }
    /**
     * @param {?} assigner
     * @return {?}
     */
    BasicDrawerService.prototype.setPropsAssigner = /**
     * @param {?} assigner
     * @return {?}
     */
    function (assigner) {
        this._propsAssigner = assigner;
    };
    return BasicDrawerService;
}());
/**
 * @abstract
 */
export { BasicDrawerService };
if (false) {
    /**
     * @type {?}
     * @protected
     */
    BasicDrawerService.prototype._propsAssigner;
    /**
     * @abstract
     * @param {?} cesiumProps
     * @param {...?} args
     * @return {?}
     */
    BasicDrawerService.prototype.add = function (cesiumProps, args) { };
    /**
     * @abstract
     * @param {?} primitive
     * @param {?} cesiumProps
     * @param {...?} args
     * @return {?}
     */
    BasicDrawerService.prototype.update = function (primitive, cesiumProps, args) { };
    /**
     * @abstract
     * @param {?} primitive
     * @return {?}
     */
    BasicDrawerService.prototype.remove = function (primitive) { };
    /**
     * @abstract
     * @return {?}
     */
    BasicDrawerService.prototype.removeAll = function () { };
    /**
     * @abstract
     * @param {?} showValue
     * @return {?}
     */
    BasicDrawerService.prototype.setShow = function (showValue) { };
    /**
     * @abstract
     * @param {?=} options
     * @return {?}
     */
    BasicDrawerService.prototype.init = function (options) { };
}
