// angular-cesium/esm2015/lib/angular-cesium/components/ac-map/ac-map.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { DOCUMENT } from '@angular/common';
import { Component, ElementRef, Inject, Input } from '@angular/core';
import { ConfigurationService } from '../../cesium-enhancements/ConfigurationService';
import { SceneMode } from '../../models/scene-mode.enum';
import { CameraService } from '../../services/camera/camera.service';
import { CesiumService } from '../../services/cesium/cesium.service';
import { ContextMenuService } from '../../services/context-menu/context-menu.service';
import { CoordinateConverter } from '../../services/coordinate-converter/coordinate-converter.service';
import { ArcDrawerService } from '../../services/drawers/arc-drawer/arc-drawer.service';
import { BillboardDrawerService } from '../../services/drawers/billboard-drawer/billboard-drawer.service';
import { CzmlDrawerService } from '../../services/drawers/czml-drawer/czml-drawer.service';
import { EllipseDrawerService } from '../../services/drawers/ellipse-drawer/ellipse-drawer.service';
import { LabelDrawerService } from '../../services/drawers/label-drawer/label-drawer.service';
import { PointDrawerService } from '../../services/drawers/point-drawer/point-drawer.service';
import { PolygonDrawerService } from '../../services/drawers/polygon-drawer/polygon-drawer.service';
import { PolylineDrawerService } from '../../services/drawers/polyline-drawer/polyline-drawer.service';
import { PolylinePrimitiveDrawerService } from '../../services/drawers/polyline-primitive-drawer/polyline-primitive-drawer.service';
import { KeyboardControlService } from '../../services/keyboard-control/keyboard-control.service';
import { CesiumEventBuilder } from '../../services/map-events-mananger/cesium-event-builder';
import { MapEventsManagerService } from '../../services/map-events-mananger/map-events-manager';
import { MapLayersService } from '../../services/map-layers/map-layers.service';
import { MapsManagerService } from '../../services/maps-manager/maps-manager.service';
import { PlonterService } from '../../services/plonter/plonter.service';
import { ScreenshotService } from '../../services/screenshot/screenshot.service';
/**
 * This is a map implementation, creates the cesium map.
 * Every layer should be tag inside ac-map tag
 *
 * Accessing cesium viewer:
 * 1. acMapComponent.getCesiumViewer()
 * 2. Use MapManagerService.getMap().getCesiumViewer() or if more then one map: MapManagerService.getMap(mapId).getCesiumViewer()
 *
 *
 * \@example
 * <ac-map>
 *     <ac-map-layer-provider></ac-map-layer-provider>
 *     <dynamic-ellipse-layer #layer></dynamic-ellipse-layer>
 * </ac-map>
 */
export class AcMapComponent {
    /**
     * @param {?} _cesiumService
     * @param {?} _cameraService
     * @param {?} _elemRef
     * @param {?} document
     * @param {?} mapsManagerService
     * @param {?} billboardDrawerService
     * @param {?} labelDrawerService
     * @param {?} ellipseDrawerService
     * @param {?} polylineDrawerService
     * @param {?} polygonDrawerService
     * @param {?} arcDrawerService
     * @param {?} pointDrawerService
     * @param {?} czmlDrawerService
     * @param {?} mapEventsManager
     * @param {?} keyboardControlService
     * @param {?} mapLayersService
     * @param {?} configurationService
     * @param {?} screenshotService
     * @param {?} contextMenuService
     * @param {?} coordinateConverter
     */
    constructor(_cesiumService, _cameraService, _elemRef, document, mapsManagerService, billboardDrawerService, labelDrawerService, ellipseDrawerService, polylineDrawerService, polygonDrawerService, arcDrawerService, pointDrawerService, czmlDrawerService, mapEventsManager, keyboardControlService, mapLayersService, configurationService, screenshotService, contextMenuService, coordinateConverter) {
        this._cesiumService = _cesiumService;
        this._cameraService = _cameraService;
        this._elemRef = _elemRef;
        this.document = document;
        this.mapsManagerService = mapsManagerService;
        this.billboardDrawerService = billboardDrawerService;
        this.labelDrawerService = labelDrawerService;
        this.ellipseDrawerService = ellipseDrawerService;
        this.polylineDrawerService = polylineDrawerService;
        this.polygonDrawerService = polygonDrawerService;
        this.arcDrawerService = arcDrawerService;
        this.pointDrawerService = pointDrawerService;
        this.czmlDrawerService = czmlDrawerService;
        this.mapEventsManager = mapEventsManager;
        this.keyboardControlService = keyboardControlService;
        this.mapLayersService = mapLayersService;
        this.configurationService = configurationService;
        this.screenshotService = screenshotService;
        this.contextMenuService = contextMenuService;
        this.coordinateConverter = coordinateConverter;
        /**
         * Disable default plonter context menu
         */
        this.disableDefaultPlonter = false;
        this.mapContainer = this.document.createElement('div');
        this.mapContainer.style.width = '100%';
        this.mapContainer.style.height = '100%';
        this.mapContainer.className = 'map-container';
        this._cesiumService.init(this.mapContainer, this);
        this._cameraService.init(this._cesiumService);
        this.mapEventsManager.init();
        this.billboardDrawerService.init();
        this.labelDrawerService.init();
        this.ellipseDrawerService.init();
        this.polylineDrawerService.init();
        this.polygonDrawerService.init();
        this.arcDrawerService.init();
        this.pointDrawerService.init();
        this.czmlDrawerService.init();
        this.keyboardControlService.init();
        this.contextMenuService.init(this.mapEventsManager);
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.mapId = this.mapsManagerService._registerMap(this.mapId, this);
        if (!this.containerId) {
            this._elemRef.nativeElement.appendChild(this.mapContainer);
        }
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
        if (changes['sceneMode']) {
            this._cameraService.setSceneMode(changes['sceneMode'].currentValue);
        }
        if (changes['flyTo']) {
            this._cameraService.cameraFlyTo(changes['flyTo'].currentValue);
        }
        if (changes['containerId'] && !changes['containerId'].firstChange) {
            /** @type {?} */
            const element = this.document.getElementById(changes['containerId'].currentValue);
            if (element) {
                element.appendChild(this.mapContainer);
            }
            else {
                throw new Error(`No element found with id: ${changes['containerId'].currentValue}`);
            }
        }
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        this.mapLayersService.drawAllLayers();
        if (this.containerId) {
            setTimeout(() => {
                /** @type {?} */
                const element = this.document.getElementById(this.containerId);
                if (element) {
                    element.appendChild(this.mapContainer);
                }
                else {
                    throw new Error(`No element found with id: ${this.containerId}`);
                }
            }, 0);
        }
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this.mapContainer.remove();
        this.mapsManagerService._removeMapById(this.mapId);
    }
    /**
     * @return {?} ac-map's cesium service
     */
    getCesiumService() {
        return this._cesiumService;
    }
    /**
     * @return {?} map's cesium viewer
     */
    getCesiumViewer() {
        return this._cesiumService.getViewer();
    }
    /**
     * @return {?}
     */
    getCameraService() {
        return this._cameraService;
    }
    /**
     * @return {?}
     */
    getId() {
        return this.mapId;
    }
    /**
     * @return {?}
     */
    getMapContainer() {
        return this.mapContainer;
    }
    /**
     * @return {?}
     */
    getMapEventsManager() {
        return this.mapEventsManager;
    }
    /**
     * @return {?}
     */
    getContextMenuService() {
        return this.contextMenuService;
    }
    /**
     * @return {?}
     */
    getScreenshotService() {
        return this.screenshotService;
    }
    /**
     * @return {?}
     */
    getKeyboardControlService() {
        return this.keyboardControlService;
    }
    /**
     * @return {?}
     */
    getCoordinateConverter() {
        return this.coordinateConverter;
    }
}
AcMapComponent.decorators = [
    { type: Component, args: [{
                selector: 'ac-map',
                template: `
    <ac-default-plonter *ngIf="!disableDefaultPlonter"></ac-default-plonter>
    <ac-context-menu-wrapper></ac-context-menu-wrapper>
    <ng-content></ng-content>
  `,
                providers: [
                    CesiumService,
                    BillboardDrawerService,
                    CesiumEventBuilder,
                    KeyboardControlService,
                    MapEventsManagerService,
                    PlonterService,
                    LabelDrawerService,
                    PolylineDrawerService,
                    PolylinePrimitiveDrawerService,
                    EllipseDrawerService,
                    PointDrawerService,
                    ArcDrawerService,
                    CzmlDrawerService,
                    PolygonDrawerService,
                    MapLayersService,
                    CameraService,
                    ScreenshotService,
                    ContextMenuService,
                    CoordinateConverter,
                ]
            }] }
];
/** @nocollapse */
AcMapComponent.ctorParameters = () => [
    { type: CesiumService },
    { type: CameraService },
    { type: ElementRef },
    { type: undefined, decorators: [{ type: Inject, args: [DOCUMENT,] }] },
    { type: MapsManagerService },
    { type: BillboardDrawerService },
    { type: LabelDrawerService },
    { type: EllipseDrawerService },
    { type: PolylineDrawerService },
    { type: PolygonDrawerService },
    { type: ArcDrawerService },
    { type: PointDrawerService },
    { type: CzmlDrawerService },
    { type: MapEventsManagerService },
    { type: KeyboardControlService },
    { type: MapLayersService },
    { type: ConfigurationService },
    { type: ScreenshotService },
    { type: ContextMenuService },
    { type: CoordinateConverter }
];
AcMapComponent.propDecorators = {
    disableDefaultPlonter: [{ type: Input }],
    mapId: [{ type: Input }],
    flyTo: [{ type: Input }],
    sceneMode: [{ type: Input }],
    containerId: [{ type: Input }]
};
if (false) {
    /**
     * Disable default plonter context menu
     * @type {?}
     */
    AcMapComponent.prototype.disableDefaultPlonter;
    /**
     * Set the id name of the map
     * default: 'default-map-id-[index]'
     * @type {?}
     */
    AcMapComponent.prototype.mapId;
    /**
     * flyTo options according to https://cesiumjs.org/Cesium/Build/Documentation/Camera.html?classFilter=cam#flyTo
     * @type {?}
     */
    AcMapComponent.prototype.flyTo;
    /**
     * Sets the map's SceneMode
     * @type {?}
     */
    AcMapComponent.prototype.sceneMode;
    /**
     * Optional - the container element's id in which the map's canvas will be appended to.
     * If not supplied - the container element will be the parent element of ac-map;
     * @type {?}
     */
    AcMapComponent.prototype.containerId;
    /**
     * @type {?}
     * @private
     */
    AcMapComponent.prototype.mapContainer;
    /**
     * @type {?}
     * @private
     */
    AcMapComponent.prototype._cesiumService;
    /**
     * @type {?}
     * @private
     */
    AcMapComponent.prototype._cameraService;
    /**
     * @type {?}
     * @private
     */
    AcMapComponent.prototype._elemRef;
    /**
     * @type {?}
     * @private
     */
    AcMapComponent.prototype.document;
    /**
     * @type {?}
     * @private
     */
    AcMapComponent.prototype.mapsManagerService;
    /**
     * @type {?}
     * @private
     */
    AcMapComponent.prototype.billboardDrawerService;
    /**
     * @type {?}
     * @private
     */
    AcMapComponent.prototype.labelDrawerService;
    /**
     * @type {?}
     * @private
     */
    AcMapComponent.prototype.ellipseDrawerService;
    /**
     * @type {?}
     * @private
     */
    AcMapComponent.prototype.polylineDrawerService;
    /**
     * @type {?}
     * @private
     */
    AcMapComponent.prototype.polygonDrawerService;
    /**
     * @type {?}
     * @private
     */
    AcMapComponent.prototype.arcDrawerService;
    /**
     * @type {?}
     * @private
     */
    AcMapComponent.prototype.pointDrawerService;
    /**
     * @type {?}
     * @private
     */
    AcMapComponent.prototype.czmlDrawerService;
    /**
     * @type {?}
     * @private
     */
    AcMapComponent.prototype.mapEventsManager;
    /**
     * @type {?}
     * @private
     */
    AcMapComponent.prototype.keyboardControlService;
    /**
     * @type {?}
     * @private
     */
    AcMapComponent.prototype.mapLayersService;
    /**
     * @type {?}
     * @private
     */
    AcMapComponent.prototype.configurationService;
    /**
     * @type {?}
     * @private
     */
    AcMapComponent.prototype.screenshotService;
    /** @type {?} */
    AcMapComponent.prototype.contextMenuService;
    /**
     * @type {?}
     * @private
     */
    AcMapComponent.prototype.coordinateConverter;
}
