// angular-cesium/esm5/lib/angular-cesium-widgets/components/toolbar/ac-toolbar/ac-toolbar.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, ElementRef, Input, Output, EventEmitter } from '@angular/core';
import { fromEvent as observableFromEvent } from 'rxjs';
import { switchMap, takeUntil, tap } from 'rxjs/operators';
import { CesiumService } from '../../../../angular-cesium/services/cesium/cesium.service';
/**
 * Toolbar widget, act as a container for ac-toolbar-button components
 * allowing drag configuration and passing `toolbarClass` as attributes
 *
 * Usage:
 * ```
 * <ac-toolbar [allowDrag]="true" (onDrag)="handleDrag($event)">
 * <ac-toolbar-button [iconUrl]="'assets/home-icon.svg'" (onClick)="goHome()">
 * </ac-toolbar-button>
 * <ac-toolbar-button [iconUrl]="'assets/explore-icon.svg'" (onClick)="rangeAndBearing()">
 * </ac-toolbar-button>
 * </ac-toolbar>
 * ```
 *
 */
var AcToolbarComponent = /** @class */ (function () {
    function AcToolbarComponent(element, cesiumService) {
        this.element = element;
        this.cesiumService = cesiumService;
        this.allowDrag = true;
        this.onDrag = new EventEmitter();
        this.dragStyle = {
            'height.px': 20,
            'width.px': 20,
        };
    }
    /**
     * @return {?}
     */
    AcToolbarComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.cesiumService.getMap().getMapContainer().appendChild(this.element.nativeElement);
        if (this.allowDrag) {
            this.listenForDragging();
        }
    };
    /**
     * @param {?} changes
     * @return {?}
     */
    AcToolbarComponent.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        if (changes.allowDrag && changes.allowDrag.currentValue && !changes.allowDrag.previousValue) {
            this.listenForDragging();
        }
        if (changes.allowDrag && !changes.allowDrag.currentValue && changes.allowDrag.previousValue) {
            this.dragSubscription.unsubscribe();
        }
    };
    /**
     * @return {?}
     */
    AcToolbarComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        if (this.dragSubscription) {
            this.dragSubscription.unsubscribe();
        }
    };
    /**
     * @private
     * @return {?}
     */
    AcToolbarComponent.prototype.listenForDragging = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        this.mouseDown$ = this.mouseDown$ || observableFromEvent(this.element.nativeElement, 'mousedown');
        this.mouseMove$ = this.mouseMove$ || (/** @type {?} */ (observableFromEvent(document, 'mousemove')));
        this.mouseUp$ = this.mouseUp$ || (/** @type {?} */ (observableFromEvent(document, 'mouseup')));
        this.drag$ = this.drag$ ||
            this.mouseDown$.pipe(switchMap(function () { return _this.mouseMove$.pipe(tap(_this.onDrag.emit), takeUntil(_this.mouseUp$)); }));
        this.dragSubscription = this.drag$.subscribe(function (event) {
            _this.element.nativeElement.style.left = event.x + "px";
            _this.element.nativeElement.style.top = event.y + "px";
        });
    };
    AcToolbarComponent.decorators = [
        { type: Component, args: [{
                    selector: 'ac-toolbar',
                    template: "\n        <div class=\"{{toolbarClass}}\">\n            <div *ngIf=\"allowDrag\">\n                <ac-toolbar-button>\n                    <ac-drag-icon></ac-drag-icon>\n                </ac-toolbar-button>\n            </div>\n\n            <ng-content></ng-content>\n        </div>\n    ",
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    styles: ["\n        :host {\n            position: absolute;\n            top: 20px;\n            left: 20px;\n            width: 20px;\n            height: 20px;\n            z-index: 999;\n            -webkit-user-drag: none;\n        }\n    "]
                }] }
    ];
    /** @nocollapse */
    AcToolbarComponent.ctorParameters = function () { return [
        { type: ElementRef },
        { type: CesiumService }
    ]; };
    AcToolbarComponent.propDecorators = {
        toolbarClass: [{ type: Input }],
        allowDrag: [{ type: Input }],
        onDrag: [{ type: Output }]
    };
    return AcToolbarComponent;
}());
export { AcToolbarComponent };
if (false) {
    /** @type {?} */
    AcToolbarComponent.prototype.toolbarClass;
    /** @type {?} */
    AcToolbarComponent.prototype.allowDrag;
    /** @type {?} */
    AcToolbarComponent.prototype.onDrag;
    /** @type {?} */
    AcToolbarComponent.prototype.dragStyle;
    /**
     * @type {?}
     * @private
     */
    AcToolbarComponent.prototype.mouseDown$;
    /**
     * @type {?}
     * @private
     */
    AcToolbarComponent.prototype.mouseMove$;
    /**
     * @type {?}
     * @private
     */
    AcToolbarComponent.prototype.mouseUp$;
    /**
     * @type {?}
     * @private
     */
    AcToolbarComponent.prototype.drag$;
    /**
     * @type {?}
     * @private
     */
    AcToolbarComponent.prototype.dragSubscription;
    /**
     * @type {?}
     * @private
     */
    AcToolbarComponent.prototype.element;
    /**
     * @type {?}
     * @private
     */
    AcToolbarComponent.prototype.cesiumService;
}
